<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Solomon Chak">
<meta name="dcterms.date" content="2024-05-06">

<title>Big Red Statistics in R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="BigRedStatisticsInR_files/libs/clipboard/clipboard.min.js"></script>
<script src="BigRedStatisticsInR_files/libs/quarto-html/quarto.js"></script>
<script src="BigRedStatisticsInR_files/libs/quarto-html/popper.min.js"></script>
<script src="BigRedStatisticsInR_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="BigRedStatisticsInR_files/libs/quarto-html/anchor.min.js"></script>
<link href="BigRedStatisticsInR_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="BigRedStatisticsInR_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="BigRedStatisticsInR_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="BigRedStatisticsInR_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="BigRedStatisticsInR_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#overview-of-the-scientific-method" id="toc-overview-of-the-scientific-method" class="nav-link active" data-scroll-target="#overview-of-the-scientific-method"><span class="header-section-number">1</span> Overview of the scientific method</a></li>
  <li><a href="#the-importance-of-statistical-tests" id="toc-the-importance-of-statistical-tests" class="nav-link" data-scroll-target="#the-importance-of-statistical-tests"><span class="header-section-number">2</span> The importance of statistical tests</a></li>
  <li><a href="#displaying-results-with-a-graph" id="toc-displaying-results-with-a-graph" class="nav-link" data-scroll-target="#displaying-results-with-a-graph"><span class="header-section-number">3</span> Displaying results with a graph</a></li>
  <li><a href="#descriptive-statistics" id="toc-descriptive-statistics" class="nav-link" data-scroll-target="#descriptive-statistics"><span class="header-section-number">4</span> Descriptive statistics</a></li>
  <li><a href="#start-using-r" id="toc-start-using-r" class="nav-link" data-scroll-target="#start-using-r"><span class="header-section-number">5</span> Start using R</a>
  <ul class="collapse">
  <li><a href="#using-r-in-rstudio-online-or-locally" id="toc-using-r-in-rstudio-online-or-locally" class="nav-link" data-scroll-target="#using-r-in-rstudio-online-or-locally"><span class="header-section-number">5.1</span> Using R in RStudio online or locally</a>
  <ul class="collapse">
  <li><a href="#installation-or-sign-up" id="toc-installation-or-sign-up" class="nav-link" data-scroll-target="#installation-or-sign-up"><span class="header-section-number">5.1.1</span> Installation or sign up</a></li>
  <li><a href="#project-setup-and-file-transfer" id="toc-project-setup-and-file-transfer" class="nav-link" data-scroll-target="#project-setup-and-file-transfer"><span class="header-section-number">5.1.2</span> Project setup and file transfer</a></li>
  <li><a href="#the-rstudio-interface" id="toc-the-rstudio-interface" class="nav-link" data-scroll-target="#the-rstudio-interface"><span class="header-section-number">5.1.3</span> The RStudio Interface</a></li>
  <li><a href="#using-packages" id="toc-using-packages" class="nav-link" data-scroll-target="#using-packages"><span class="header-section-number">5.1.4</span> Using packages</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#importing-data-into-r" id="toc-importing-data-into-r" class="nav-link" data-scroll-target="#importing-data-into-r"><span class="header-section-number">6</span> Importing data into R</a>
  <ul class="collapse">
  <li><a href="#preparing-a-.csv-file-from-excel" id="toc-preparing-a-.csv-file-from-excel" class="nav-link" data-scroll-target="#preparing-a-.csv-file-from-excel"><span class="header-section-number">6.1</span> Preparing a .csv file from Excel</a></li>
  <li><a href="#loading-a-.csv-file-into-r" id="toc-loading-a-.csv-file-into-r" class="nav-link" data-scroll-target="#loading-a-.csv-file-into-r"><span class="header-section-number">6.2</span> Loading a .csv file into R</a></li>
  <li><a href="#loading-an-excel-file-directly" id="toc-loading-an-excel-file-directly" class="nav-link" data-scroll-target="#loading-an-excel-file-directly"><span class="header-section-number">6.3</span> Loading an Excel file directly</a></li>
  </ul></li>
  <li><a href="#displaying-results-with-a-graph-1" id="toc-displaying-results-with-a-graph-1" class="nav-link" data-scroll-target="#displaying-results-with-a-graph-1"><span class="header-section-number">7</span> Displaying results with a graph</a>
  <ul class="collapse">
  <li><a href="#how-to-construct-a-bar-graph" id="toc-how-to-construct-a-bar-graph" class="nav-link" data-scroll-target="#how-to-construct-a-bar-graph"><span class="header-section-number">7.1</span> How to Construct a Bar Graph</a></li>
  <li><a href="#how-to-construct-a-scatter-plot-with-a-linear-fit" id="toc-how-to-construct-a-scatter-plot-with-a-linear-fit" class="nav-link" data-scroll-target="#how-to-construct-a-scatter-plot-with-a-linear-fit"><span class="header-section-number">7.2</span> How to Construct a Scatter Plot with a Linear Fit</a></li>
  </ul></li>
  <li><a href="#how-to-generate-descriptive-statistics" id="toc-how-to-generate-descriptive-statistics" class="nav-link" data-scroll-target="#how-to-generate-descriptive-statistics"><span class="header-section-number">8</span> How to generate descriptive statistics</a></li>
  <li><a href="#inferential-statistics" id="toc-inferential-statistics" class="nav-link" data-scroll-target="#inferential-statistics"><span class="header-section-number">9</span> Inferential statistics</a>
  <ul class="collapse">
  <li><a href="#linear-regression" id="toc-linear-regression" class="nav-link" data-scroll-target="#linear-regression"><span class="header-section-number">9.1</span> Linear regression</a>
  <ul class="collapse">
  <li><a href="#visualization-scatter-plot" id="toc-visualization-scatter-plot" class="nav-link" data-scroll-target="#visualization-scatter-plot"><span class="header-section-number">9.1.1</span> Visualization: scatter plot</a></li>
  </ul></li>
  <li><a href="#t-test" id="toc-t-test" class="nav-link" data-scroll-target="#t-test"><span class="header-section-number">9.2</span> T-test</a>
  <ul class="collapse">
  <li><a href="#visualization-boxplot" id="toc-visualization-boxplot" class="nav-link" data-scroll-target="#visualization-boxplot"><span class="header-section-number">9.2.1</span> <strong>Visualization: boxplot</strong></a></li>
  </ul></li>
  <li><a href="#anova" id="toc-anova" class="nav-link" data-scroll-target="#anova"><span class="header-section-number">9.3</span> ANOVA</a>
  <ul class="collapse">
  <li><a href="#global-anova-test" id="toc-global-anova-test" class="nav-link" data-scroll-target="#global-anova-test"><span class="header-section-number">9.3.1</span> Global ANOVA test</a></li>
  <li><a href="#post-hoc-test" id="toc-post-hoc-test" class="nav-link" data-scroll-target="#post-hoc-test"><span class="header-section-number">9.3.2</span> <strong><em>Post hoc test</em></strong></a></li>
  <li><a href="#visualization-box-plot-and-bar-plot" id="toc-visualization-box-plot-and-bar-plot" class="nav-link" data-scroll-target="#visualization-box-plot-and-bar-plot"><span class="header-section-number">9.3.3</span> <strong>Visualization: box plot and bar plot</strong></a></li>
  </ul></li>
  <li><a href="#two-types-of-chi-square-x2-tests" id="toc-two-types-of-chi-square-x2-tests" class="nav-link" data-scroll-target="#two-types-of-chi-square-x2-tests"><span class="header-section-number">9.4</span> Two types of Chi-square (X<sup>2</sup>) tests</a>
  <ul class="collapse">
  <li><a href="#chi-square---goodness-of-fit-test" id="toc-chi-square---goodness-of-fit-test" class="nav-link" data-scroll-target="#chi-square---goodness-of-fit-test"><span class="header-section-number">9.4.1</span> Chi-square - Goodness of Fit test</a></li>
  <li><a href="#chi-square---contingency-test" id="toc-chi-square---contingency-test" class="nav-link" data-scroll-target="#chi-square---contingency-test"><span class="header-section-number">9.4.2</span> Chi-square - Contingency test</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Big Red Statistics in R</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Solomon Chak </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">May 6, 2024</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="overview-of-the-scientific-method" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Overview of the scientific method</h1>
<p>(Will have the same text as the current manual)</p>
<hr>
</section>
<section id="the-importance-of-statistical-tests" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> The importance of statistical tests</h1>
<p>(Will have the same text as the current manual)</p>
<hr>
</section>
<section id="displaying-results-with-a-graph" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Displaying results with a graph</h1>
<p>This section is moved after “Importing data into R”: <a href="#displaying-results-with-a-graph">Displaying results with a graph</a></p>
<hr>
</section>
<section id="descriptive-statistics" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Descriptive statistics</h1>
<p><strong><em>Mean</em></strong></p>
<p>The <strong>mean</strong> <span class="math inline">\(\bar{x}\)</span> (often referred to as the average) corresponds to the sum of the measurements divided by the number of measurements (as in the case of 10, 40, 25, 35, and 40, which produce a mean of 30). In the case of an odd number of measurements, the <strong>median</strong> corresponds the measurement in the middle when all of the measurements are sorted from lowest to highest (as in the case of 10, 40, 25, 35, and 40, which produces a median of 35). In the case of an even number of measurements, the median corresponds to the mean of the two measurements in the middle when all of the measurements are sorted from lowest to highest. Finally, the <strong>mode</strong> corresponds to the most common measurement (as in the case of 10, 40, 25, 35, and 40, which produces a mode of 40).</p>
<p><strong><em>Standard deviation</em></strong></p>
<p>The <strong>standard deviation</strong> (<em>s</em>) is a measure of the distribution of the data (i.e.&nbsp;extent of variation among different subjects) and weights each value by its distance from the mean.&nbsp; There is an increase in the standard deviation if there is a large distance of the values from the mean (as in the case of 10, 40, 25, 35, 40, which have a mean of 30).&nbsp; In contrast, there is a decrease in the standard deviation if there is a small distance (if any) of the values from the mean (as in the case of 25, 30, 35, 30, 30, which have the same mean of 30). The result for each group is summarized as the mean “plus or minus” the standard deviation (<span class="math inline">\(\bar{x}\)</span> ± <em>s</em>).&nbsp; This gives you a lot of information about the data set. If someone describes a data set as the mean plus or minus one SD, they are giving you the range into which 68% of the data points fall.&nbsp; Mean ± 2 SD gives you the range of 95% of the data, and Mean ± 3 SD encompasses 99% of the data.&nbsp; Note that the standard deviation is sometimes represented by vertical lines known as <strong>error bars</strong> that extend from the mean in both an upward and downward direction.</p>
<p><strong><em>Standard error</em></strong></p>
<p>The standard error of the mean, or simply the <strong>standard error</strong> (SE) is another common way to state the variation around the mean.&nbsp; Instead of telling you about the range of the data set, it tells you the range of the “true” mean.&nbsp; If you added one sample to your study or took one sample away, the mean may change a little.&nbsp; So your mean is correct for your group of samples, but not necessarily for the whole population of all organisms that could have been in your study.&nbsp; SEM tells you that if you could have included every possible subject or sample - the mean probably would have been somewhere in this range.&nbsp; Once you have a calculated the standard deviation of the mean (s), it is easy to calculate SE.</p>
<p><img src="Images/SE.png" class="img-fluid" width="197"></p>
<p>n = the number of samples or trials. This technique results in a smaller number than s offers, and thus often makes graphs with error bars look smaller. Whether to use s or SE in reporting your statistics and making error bars varies depending on the application, style and field. Be sure to find out what your instructor prefers for your assignments! And always state in your figure legends what your error bars are showing (e.g., “Bars show mean and error bars represent 1SD”).</p>
<p><img src="Images/errorbar.png" class="img-fluid"></p>
<p><strong><em>Coefficient of variation</em></strong></p>
<p>The <strong>coefficient of variation (CV)</strong> corresponds to the standard deviation divided by the mean and is useful when comparing the amount of variation for one variable among groups with different means.</p>
<p>The variance in data sometimes fits a <strong>normal distribution</strong> (i.e.&nbsp;bell curve on a graph); such data are subject to a parametric test. However, the variance in data sometimes does not fit a normal distribution; such data are subject to a non-parametric test (addressed in advanced coursework).</p>
<hr>
</section>
<section id="start-using-r" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Start using R</h1>
<section id="using-r-in-rstudio-online-or-locally" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="using-r-in-rstudio-online-or-locally"><span class="header-section-number">5.1</span> Using R in RStudio online or locally</h2>
<section id="installation-or-sign-up" class="level3" data-number="5.1.1">
<h3 data-number="5.1.1" class="anchored" data-anchor-id="installation-or-sign-up"><span class="header-section-number">5.1.1</span> Installation or sign up</h3>
<p>R is a command line program where you interact with it through text (command line) instead of a graphical user interface (e.g., Windows, Excel, etc.). There are two options for running R. Your instructor will let you know how you should use R for the class.</p>
<p>In both options, you will not interact directly with R, but through an integrated development environment called Rstudio on your computer or online as Posit.</p>
<ul>
<li><p>You can use RStudio online. Sign up for an account <a href="https://posit.co/products/cloud/cloud/">here</a>. The “Cloud Free” option will be sufficient. Do not choose the PLUS or other options. Your instructor will send you a link to get access to the class space.</p></li>
<li><p>You can also download RStudio and install it on your computer. Note that depends on the version of RStudio and R, the interface and commands could be slightly different from the online version. Follow the instructions <a href="https://posit.co/download/rstudio-desktop/">here</a> to download and install both R and RStudio.</p></li>
</ul>
</section>
<section id="project-setup-and-file-transfer" class="level3" data-number="5.1.2">
<h3 data-number="5.1.2" class="anchored" data-anchor-id="project-setup-and-file-transfer"><span class="header-section-number">5.1.2</span> Project setup and file transfer</h3>
<p><strong>Online option</strong></p>
<p>If you choose the cloud option, you will need to upload data/ download output files to/from the cloud. But the benefit is that you can use R anywhere, and on virtually any machine with internet (even a tablet)!</p>
<p>Once in the classroom space, you will see any project assignments that your instructor has created for you. Your instructor will also have access to your assignments in this space. Your instructor may also have created a template for you to work from that includes pre-installed packages.</p>
<p>To create your own project on posit.cloud, click <code>New Project &gt; New RStudio Project</code> on the upper right. Once the project is deployed, on the lower right panel, click the <code>upload</code> button and choose the data file from your computer.</p>
<p>To download a file, check the box next to the file, click the <code>More &gt; Export…</code> and click <code>Download</code>.</p>
<p><strong>Local option</strong></p>
<p>If you choose the local option, you should set up a project and a folder that is associated with this project where all files will be stored.</p>
<p>For example, create a folder on your desktop that is called “R_is_fun”. Start RStudio, click <code>File &gt; New Project…</code> and select <code>Existing Directory</code> and browse to the folder <code>R_is_fun</code>.</p>
<p>Then click <code>Create Project</code>. Now you can access all files in the folder <code>R_is_fun</code> (without indicating the full pathname). Everything you saved in R will also appear in this folder.</p>
</section>
<section id="the-rstudio-interface" class="level3" data-number="5.1.3">
<h3 data-number="5.1.3" class="anchored" data-anchor-id="the-rstudio-interface"><span class="header-section-number">5.1.3</span> The RStudio Interface</h3>
<p>Let’s introduce what is being displayed in RStudio. Before we get started, click <code>File &gt; New File &gt; R Script</code>. This will result in a default four-panel interface <a href="#fig-RStudio_interface">Figure&nbsp;1</a>.</p>
<div id="fig-RStudio_interface" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="Images/Rstudio_interface.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;1: The four-panel interface of RStudio.</figcaption>
</figure>
</div>
<ul>
<li><p>(A) The upper left is the R script panel. This is where you will write and save your R script or commands. The panel should be saved regularly, which serves as a record of all the commands that you have (or are planning) to run. For example, clicking File &gt; Save and giving it the name “R_is_fun” will save it as “R_is_fun.R” in your working directory. The scripts here are not being read by R until you click the “Run” button on the top right.</p></li>
<li><p>(B) The lower left is the R console, which is the innate command line interface of R. A command from the R script above will be displayed here when you click the “Run” button, which is when R “reads” your command. Any output or error from R will also be displayed here. You could type your command directly here, but it will not be saved in the R script.</p></li>
<li><p>(C) The upper right is the environment panel. Objects that are read or generated in R (such as dataframes) will be displayed here, along with any packages that you load and functions that you create. History provides all of your commands from the sessionConnections will enable you to connect to data sources. Tutorials will let you run R tutorials.</p></li>
<li><p>(D) The lower right is the files panel. It displays the files in your working directory. In Posit cloud, you will see files uploaded to the course space by your instructor. It also displays figures in a “Plots” tab if your commands result in graphical outputs. The Packages tab indicates which packages are currently installed, while check marks indicate which ones are loaded in the library. The Help tab displays the manual for each package and command.</p></li>
</ul>
<p>For a more comprehensive introduction to RStudio, see <a href="https://education.rstudio.com/learn/beginner/">here</a>.</p>
</section>
<section id="using-packages" class="level3" data-number="5.1.4">
<h3 data-number="5.1.4" class="anchored" data-anchor-id="using-packages"><span class="header-section-number">5.1.4</span> Using packages</h3>
<p>Many packages are available to expand the basic functions of R.</p>
<p>To use these packages, you will need to install them first. Then you can load the package as libraries.</p>
<p>We will install a few packages that will be used in this document. You may be asked to select a mirror to download the packages. <em>Note that you will only need to do this ONCE for each Project.</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"readxl"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"Rmisc"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"reshape"</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("tidyverse")</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"agricolae"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then, we will load this packages such that functions from these packages can be used. <em>Note that you WILL need to run this each time when you start your Project.</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"readxl"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"Rmisc"</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"reshape"</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co">#library("tidyverse")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
</section>
<section id="importing-data-into-r" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Importing data into R</h1>
<section id="preparing-a-.csv-file-from-excel" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="preparing-a-.csv-file-from-excel"><span class="header-section-number">6.1</span> Preparing a .csv file from Excel</h2>
<p>Your data should be entered and stored in spreadsheet programs such as Microsoft Excel before being imported into R. While R can read Excel files (A) such as .xls and .xlsx, data are typically saved into a simpler text format such as .csv (comma-separated values, B), where each row appears in a new line and each column is separated by a comma (,) (<a href="#fig-excel">Figure&nbsp;2</a>).</p>
<div id="fig-excel" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="Images/excel_vs_csv.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;2: Comparison of the same table in (A) Excel and (B) .csv format.</figcaption>
</figure>
</div>
<p>A spreadsheet in Excel can be exported into .csv by clicking <code>File &gt; Save As…</code> and selecting <code>CSV</code> in the drop-down manual under <code>File Format</code>.</p>
<p>Try doing this with <code>name.xlsx</code> which can be downloaded from <a href="https://github.com/solomonchak/R_Intro/raw/5d08edb7d4981d78066a6a5294567e735fd00350/File/name.xlsx">here</a>.</p>
<p>Place the .csv file into your project’s directory (e.g., R_is_fun) or upload it to your project in posit.cloud using the Files pane (see <a href="#project-setup-and-file-transfer">Project setup and file transfer</a>).</p>
<p>Now, you should see the .csv file listed in the lower right panel of RStudio (<strong>?@fig-file_panel</strong>).</p>
<p><img src="Images/filepanel.png" id="fig-file_panel" class="img-fluid" alt="The file panel after placing the file “name.csv” into the working directory or after uploading it to posit.cloud.">.</p>
</section>
<section id="loading-a-.csv-file-into-r" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="loading-a-.csv-file-into-r"><span class="header-section-number">6.2</span> Loading a .csv file into R</h2>
<p>Note that while the name.csv file is in the file panel, it’s not being loaded into R yet. It is merely present in the folder on the cloud or your computer. To load the file into R, in the R script panel, type:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>data_name <span class="ot">=</span> <span class="fu">read.csv</span>(<span class="at">file =</span> <span class="st">"name.csv"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># In most command-line interfaces, texts after # are not read by the software and serve as notes.</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Here, read.csv() is a function.</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co"># You can pass arguments inside the ().</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co"># A function could have no argument or many arguments.</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co"># In read.csv(), the first argument is called file and we pass "name.csv" to this argument.</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Since R knows the expected sequence of arguments (this is called positional argument), you could skip the name and just write:</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="co"># data_name = read.csv("name.csv")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>(Note that you can copy everything on the above “code chunk” using the copy button that appears when you hover your cruiser on the code chunk.)</p>
<p>What the code does is load “name.csv” into R and save it as an object called “data_name”. Now the object “data_name” should appear in the environment panel (<a href="#fig-env_panel">Figure&nbsp;3</a>).</p>
<div id="fig-env_panel" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="Images/env_panel.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;3: An object loaded into R will appear in the environmental panel.</figcaption>
</figure>
</div>
<p>In the environmental panel, you can click on the object to open it in a new tab in the R script panel. Note that doing it actually sends a script to the R console as: <code>View(data_name)</code></p>
<p>You can also view the object within the R console by calling it directly:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>data_name</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  First_name Height_ft Age Academic_standing
1     Denzel       5.8  24         Sophomore
2     Claire       5.0  20            Junior
3       Jack       7.2  29        First-year
4     Quisha       5.7  26         Sophomore
5     Carlos       4.9  24        First-year
6      Julio       5.3  28         Sophomore
7     Hiromi       6.0  26            Senior
8      Jiang       5.5  28            Junior
9       Isha       4.4  18            Senior</code></pre>
</div>
</div>
<p>You can see that the file has loaded correctly. The 9 rows and 4 columns are present, as well as the column titles.</p>
<p>The object is a <em>dataframe</em>, which is a class of R object that stores table. See <a href="https://swcarpentry.github.io/r-novice-inflammation/13-supp-data-structures/">here</a> for other classes in R.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(data_name)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data.frame"</code></pre>
</div>
</div>
<p>You can get more information of a dataframe with:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(data_name)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   9 obs. of  4 variables:
 $ First_name       : chr  "Denzel" "Claire" "Jack" "Quisha" ...
 $ Height_ft        : num  5.8 5 7.2 5.7 4.9 5.3 6 5.5 4.4
 $ Age              : int  24 20 29 26 24 28 26 28 18
 $ Academic_standing: chr  "Sophomore" "Junior" "First-year" "Sophomore" ...</code></pre>
</div>
</div>
<p>In some cases, you may need to add more arguments to <code>read.csv()</code>. Explore the documentation of this function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">help</span>(<span class="st">"read.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can also search for it in the Help tab in the file panel. Another useful tip is that after you type a command, like aov(), you can hit the tab key and RStudio will show the possible arguments. Note also that RStudio will auto-complete the closing bracket for you.</p>
</section>
<section id="loading-an-excel-file-directly" class="level2" data-number="6.3">
<h2 data-number="6.3" class="anchored" data-anchor-id="loading-an-excel-file-directly"><span class="header-section-number">6.3</span> Loading an Excel file directly</h2>
<p>You can also directly import an Excel file into R with the help of the <code>readxl</code> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"readxl"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If you saved your Excel file as <code>name.xlsx</code>, then the command below will read the first sheet of an Excel file by default.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">read_xlsx</span>(<span class="at">path =</span> <span class="st">"name.xlsx"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 9 × 4
  First_name Height_ft   Age Academic_standing
  &lt;chr&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;            
1 Denzel           5.8    24 Sophomore        
2 Claire           5      20 Junior           
3 Jack             7.2    29 First-year       
4 Quisha           5.7    26 Sophomore        
5 Carlos           4.9    24 First-year       
6 Julio            5.3    28 Sophomore        
7 Hiromi           6      26 Senior           
8 Jiang            5.5    28 Junior           
9 Isha             4.4    18 Senior           </code></pre>
</div>
</div>
<p>Note that the above code does not save or overwrite the object <code>data_name</code> because we did not do the <code>data_name = read_xlsx(path = "name.xlsx")</code>. It simply read and output the result into the R console.</p>
<hr>
</section>
</section>
<section id="displaying-results-with-a-graph-1" class="level1" data-number="7">
<h1 data-number="7"><span class="header-section-number">7</span> Displaying results with a graph</h1>
<section id="how-to-construct-a-bar-graph" class="level2" data-number="7.1">
<h2 data-number="7.1" class="anchored" data-anchor-id="how-to-construct-a-bar-graph"><span class="header-section-number">7.1</span> How to Construct a Bar Graph</h2>
<p>We will use the dataframe that we have loaded before: <code>data_name</code> and the package <code>ggplot2</code>.</p>
<p>We will plot the average <code>Height_ft</code> across different <code>Academic_standing</code> and add an error bar that represents the standard deviation (or standard error if needed).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We will first have R calculate the mean, sd, and se by group. summarySE is a function in Rmisc.</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>Height_by_Academic_summary <span class="ot">=</span> <span class="fu">summarySE</span>(<span class="at">data =</span> data_name, <span class="at">measurevar =</span> <span class="st">"Height_ft"</span>, <span class="at">groupvars =</span> <span class="st">"Academic_standing"</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at Height_by_Academic_summary</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>Height_by_Academic_summary</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Academic_standing N Height_ft        sd        se         ci
1        First-year 2      6.05 1.6263456 1.1500000 14.6121354
2            Junior 2      5.25 0.3535534 0.2500000  3.1765512
3            Senior 2      5.20 1.1313708 0.8000000 10.1649638
4         Sophomore 3      5.60 0.2645751 0.1527525  0.6572411</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We will also convert Height_ft to be an ordered factor so that it will plot accordingly to the levels specified here, instead of alphabetically </span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>Height_by_Academic_summary<span class="sc">$</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>Academic_standing_order <span class="ot">=</span> <span class="fu">factor</span>(Height_by_Academic_summary<span class="sc">$</span>Academic_standing, <span class="at">ordered=</span>T, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"First-year"</span>, <span class="st">"Sophomore"</span>, <span class="st">"Junior"</span>, <span class="st">"Senior"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can then use Height_by_Academic_summary to build a barplot</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> Height_by_Academic_summary, <span class="fu">aes</span>(<span class="at">x =</span> Academic_standing_order, <span class="at">y =</span> Height_ft)) <span class="sc">+</span> <span class="co"># This line is only setting up the dataset and aesthetic mappings</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>) <span class="sc">+</span> <span class="co"># This line adds the layer of bar plot</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin=</span>Height_ft<span class="sc">-</span>sd, <span class="at">ymax=</span>Height_ft<span class="sc">+</span>sd), <span class="at">width=</span>.<span class="dv">2</span>,</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>                 <span class="at">position=</span><span class="fu">position_dodge</span>(.<span class="dv">9</span>)) <span class="sc">+</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>   <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Academic standing"</span>, <span class="at">y =</span> <span class="st">"Height (feet)"</span>)  <span class="co"># The last two lines are additional layers that add the error bar and edit the axis labels.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># You will need at least the firs two lines (connected by +) to build a ggplot:</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot(data = Height_by_Academic_summary, aes(x = Academic_standing_order, y = Height_ft)) + </span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="co"># geom_bar(stat="identity") </span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="co"># You could save the first two lines as an object and add additional layers:</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co"># p = ggplot(data = Height_by_Academic_summary, aes(x = Academic_standing_order, y = Height_ft)) + geom_bar(stat="identity") </span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="co"># p + geom_errorbar(aes(ymin=Height_ft-sd, ymax=Height_ft+sd), width=.2, position=position_dodge(.9)) +  labs(x = "Academic standing", y = "Height (feet)") </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can also overlap the data points on the bar graph.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># It requires getting the raw data from data_name</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> Height_by_Academic_summary, <span class="fu">aes</span>(<span class="at">x =</span> Academic_standing_order, <span class="at">y =</span> Height_ft)) <span class="sc">+</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>) <span class="sc">+</span> </span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin=</span>Height_ft<span class="sc">-</span>sd, <span class="at">ymax=</span>Height_ft<span class="sc">+</span>sd), <span class="at">width=</span>.<span class="dv">2</span>,</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>                 <span class="at">position=</span><span class="fu">position_dodge</span>(.<span class="dv">9</span>)) <span class="sc">+</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>   <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Academic standing"</span>, <span class="at">y =</span> <span class="st">"Height (feet)"</span>)  <span class="sc">+</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> data_name, <span class="fu">aes</span>(<span class="at">x =</span> Academic_standing, <span class="at">y =</span> Height_ft), <span class="at">position =</span> <span class="st">"jitter"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Another good way to display the same data is to use a boxplot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> data_name, <span class="fu">aes</span>(<span class="at">x =</span> Academic_standing, <span class="at">y =</span> Height_ft)) <span class="sc">+</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">outlier.colour =</span> <span class="st">"grey"</span>) <span class="sc">+</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">width =</span> <span class="fl">0.1</span>, <span class="at">height =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="how-to-construct-a-scatter-plot-with-a-linear-fit" class="level2" data-number="7.2">
<h2 data-number="7.2" class="anchored" data-anchor-id="how-to-construct-a-scatter-plot-with-a-linear-fit"><span class="header-section-number">7.2</span> How to Construct a Scatter Plot with a Linear Fit</h2>
<p>We will use the dataframe that we have loaded before: <code>data_name</code> and the package <code>ggplot2</code>.</p>
<p>We will plot <code>Age</code> as the predictor (independent) variable on the X axis and <code>Height_ft</code> as the response (dependent) variable on the Y axis. A linear regression line will be added.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> data_name, <span class="fu">aes</span>(<span class="at">x =</span> Age, Height_ft)) <span class="sc">+</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">se =</span> F, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span> </span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Height (feet)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<hr>
</section>
</section>
<section id="how-to-generate-descriptive-statistics" class="level1" data-number="8">
<h1 data-number="8"><span class="header-section-number">8</span> How to generate descriptive statistics</h1>
<p>Let’s say that you measure the length of each of 12 seashells. Here are the instructions for determining the mean and standard deviation using R:</p>
<p>Follow the instructions in the previous section to enter your continuous data into a new dataframe called <code>seashells</code>. In this case your dependent variable is <code>Length_cm</code> of the seashells.</p>
<p>For example, you can first enter your data into Excel (<strong>?@fig-seashellexcel</strong>), and save if as <code>seashells.csv</code> in your project’s directory.</p>
<p><img src="Images/seashell%20data%20in%20excel.png" id="fig-seashellexcel" class="img-fluid" width="188" alt="Entering the data in Excel.">.</p>
<p>Then you can run this to load it as a dataframe called <code>seashells</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>seashells <span class="ot">=</span> <span class="fu">read.csv</span>(<span class="st">"seashells.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Alternative, you also run this code to manually enter the data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>seashells <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="st">"Length_cm"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="fl">4.5</span>, <span class="fl">3.1</span>, <span class="fl">4.2</span>, <span class="fl">4.4</span>, <span class="dv">4</span>, <span class="fl">3.6</span>, <span class="fl">3.7</span>, <span class="fl">3.1</span>, <span class="fl">2.8</span>, <span class="fl">4.3</span>, <span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Either way, your data should look like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>seashells</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Length_cm
1        3.0
2        4.5
3        3.1
4        4.2
5        4.4
6        4.0
7        3.6
8        3.7
9        3.1
10       2.8
11       4.3
12       5.0</code></pre>
</div>
</div>
<p>You can check whether Length_cm is continuous (i.e, numeric).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(seashells<span class="sc">$</span>Length_cm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "numeric"</code></pre>
</div>
</div>
<p>Generate basic summary statistics.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(seashells<span class="sc">$</span>Length_cm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  2.800   3.100   3.850   3.808   4.325   5.000 </code></pre>
</div>
</div>
<p>Generate standard deviation (SD).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sd</span>(seashells<span class="sc">$</span>Length_cm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.7012435</code></pre>
</div>
</div>
<p>Generate standard error (SE) using the formula</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sd</span>(seashells<span class="sc">$</span>Length_cm)<span class="sc">/</span><span class="fu">sqrt</span>(<span class="fu">length</span>(seashells<span class="sc">$</span>Length_cm)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.2024316</code></pre>
</div>
</div>
<p>Generate standard deviation (SD) and standard error (SE) using the package Rmisc</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"Rmisc"</span>) <span class="co"># You don't need to write this everytime if you have loaded the library before.</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summarySE</span>(seashells, <span class="st">"Length_cm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   .id  N Length_cm        sd        se        ci
1 &lt;NA&gt; 12  3.808333 0.7012435 0.2024316 0.4455489</code></pre>
</div>
</div>
<p>You can easily calculate the means and standard deviations of different groups in R at the same time. Let’s say you measured 12 shells again, but this experiment you used 3 different species. You can calculate the means and standard deviations of the different species by inputting your data in the following way:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We will add another column to the dataframe</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>seashells<span class="sc">$</span>Species <span class="ot">=</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"Species1"</span>, <span class="dv">4</span>), <span class="fu">rep</span>(<span class="st">"Species2"</span>, <span class="dv">4</span>), <span class="fu">rep</span>(<span class="st">"Species3"</span>, <span class="dv">4</span>))</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="co"># To view the data:</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>seashells</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Length_cm  Species
1        3.0 Species1
2        4.5 Species1
3        3.1 Species1
4        4.2 Species1
5        4.4 Species2
6        4.0 Species2
7        3.6 Species2
8        3.7 Species2
9        3.1 Species3
10       2.8 Species3
11       4.3 Species3
12       5.0 Species3</code></pre>
</div>
</div>
<p>Then use <code>summarySE</code> from <code>Rmisc</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summarySE</span>(<span class="at">data =</span> seashells, <span class="at">measurevar =</span> <span class="st">"Length_cm"</span>, <span class="at">groupvars =</span> <span class="st">"Species"</span>, )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Species N Length_cm        sd        se        ci
1 Species1 4     3.700 0.7615773 0.3807887 1.2118394
2 Species2 4     3.925 0.3593976 0.1796988 0.5718819
3 Species3 4     3.800 1.0295630 0.5147815 1.6382645</code></pre>
</div>
</div>
<p>With this data, you can make a table to summarize descriptive statistics of seashell lengths from three different species (<a href="#fig-meanshell">Figure&nbsp;4</a>).</p>
<div id="fig-meanshell" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="Images/Mean%20shell%20length%20by%20species.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;4: Mean shell length ( 1 standard deviation) for species 1, 2, and 3.</figcaption>
</figure>
</div>
<hr>
</section>
<section id="inferential-statistics" class="level1" data-number="9">
<h1 data-number="9"><span class="header-section-number">9</span> Inferential statistics</h1>
<p>Step One: Figure out what statistic you should use! How do you do that? First, you should be able to clearly write out your experimental question, your hypothesis, and identify your dependent and independent variables. Go ahead – do it now – I’ll wait.</p>
<p>Once you have your variables identified, determine if they are continuous (measurement) data or nominal (categorical) data. Now you can use the chart below to figure out which type of statistic is right for you (<a href="#fig-stat">Figure&nbsp;5</a>). Note that each combination of variable types also has a type of graph or table that is typically used to display the data! Instructions for how to run each stat and make each figure are given in the sections that follow.</p>
<div id="fig-stat" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="Images/Statistic_Chart.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;5: Statistic Decision Chart.</figcaption>
</figure>
</div>
<section id="linear-regression" class="level2" data-number="9.1">
<h2 data-number="9.1" class="anchored" data-anchor-id="linear-regression"><span class="header-section-number">9.1</span> Linear regression</h2>
<p>A linear regression is employed to determine if there is a relationship between two variables, X and Y <u><strong>when both variables are continuous</strong></u> (measurement) type data. A p value that is less than 0.05 indicates that there is less than 5% probability that the association you observe is due to chance under the null hypothesis.&nbsp; In practice this means you can reject the null hypothesis and conclude that something other than chance is producing the relationship that you observe. This statistical test will also indicate the direction and magnitude of the relationship. The direction is inferred from the slope of the line (m), while the magnitude or strength of the relationship is inferred from the coefficient of determination (R<sup>2</sup>). If R<sup>2</sup> = 1, then the line is a perfect “fit,” meaning that Y can be predicted from X (or vice versa) using the equation of the line (y = mx + b). If R<sup>2</sup> &lt; 1, then the line is not a perfect “fit,” meaning that only a proportion of Y can be predicted from X (or vice versa) using the equation of the line (y = mx + b). For example, if R<sup>2</sup> = 0.94, then 94% of the variation in Y can be explained by X. Note that the R<sup>2</sup> and line equation are ONLY relevant if there is a significant relationship (p &lt; 0.05). If there is no significant relationship, then the line and its equation are arbitrary and not meaningful. Likewise, you can’t describe the magnitude of a relationship (R<sup>2</sup>) if there is no relationship.</p>
<p>Let’s say that want to determine if temperature is related to chirp rates of the snowy tree cricket. You measure the temperature and the rate of 10 chirping crickets at those respective temperatures. Since crickets are ectoderms and their metabolism is effected by their external environment, we should expect to see an increase in chirping as temperature increases.</p>
<p>Load the data into R by running the code below. We have two continuous variables <code>temperature</code> and <code>chips_per_min</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>chirps <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">temperature =</span> <span class="fu">c</span>(<span class="dv">31</span>, <span class="dv">22</span>, <span class="dv">34</span>, <span class="dv">29</span>, <span class="dv">27</span>, <span class="dv">24</span>, <span class="dv">21</span>, <span class="dv">29</span>, <span class="dv">26</span>, <span class="dv">29</span>), <span class="at">chirps_per_min =</span> <span class="fu">c</span>(<span class="dv">79</span>, <span class="dv">64</span>, <span class="dv">79</span>, <span class="dv">73</span>, <span class="dv">69</span>, <span class="dv">62</span>, <span class="dv">59</span>, <span class="dv">70</span>, <span class="dv">64</span>, <span class="dv">68</span>))</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>chirps</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   temperature chirps_per_min
1           31             79
2           22             64
3           34             79
4           29             73
5           27             69
6           24             62
7           21             59
8           29             70
9           26             64
10          29             68</code></pre>
</div>
</div>
<p>In R the <code>lm</code> function can fit a number of linear models, including linear regression.</p>
<p>The code below creates a linear model (lm_HA) to predict <code>chirps_per_min</code> (the response or dependent variable) by <code>temperature</code> (the predictor or independent variable):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>lm_CT <span class="ot">=</span> <span class="fu">lm</span>(<span class="at">formula =</span> chirps_per_min <span class="sc">~</span> temperature, <span class="at">data =</span> chirps)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>lm_CT</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = chirps_per_min ~ temperature, data = chirps)

Coefficients:
(Intercept)  temperature  
     26.389        1.556  </code></pre>
</div>
</div>
<p>Calling lm_CT shows the coefficients for intercept and slope (under temperature) of the linear regression. But we can get much more out of lm_CT using <code>summary()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(lm_CT)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = chirps_per_min ~ temperature, data = chirps)

Residuals:
    Min      1Q  Median      3Q     Max 
-3.5000 -1.6667 -0.1667  1.2778  4.3889 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)  26.3889     6.1375   4.300 0.002617 ** 
temperature   1.5556     0.2234   6.962 0.000117 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 2.714 on 8 degrees of freedom
Multiple R-squared:  0.8583,    Adjusted R-squared:  0.8406 
F-statistic: 48.47 on 1 and 8 DF,  p-value: 0.000117</code></pre>
</div>
</div>
<p>Under “Coefficients”, it shows a t-test of whether the intercept and slope (temperature) is significantly different from zero.</p>
<p>In the last three lines, it shows:</p>
<ul>
<li><p>Residual standard error: lower values suggest better fit (i.e., less distance between the data point and the predicted line).</p></li>
<li><p>Degrees of freedom: the maximum number of logically independent values.</p></li>
<li><p>Multiple R-squared: the percentage of the response variable variation that a the model explains.</p></li>
<li><p>Adjusted R-squared: R-squared adjusted for the number of terms in the model. Use this to compare between models with different number of predictor variables.</p></li>
<li><p>Statistics of a F-test that evaluates the fit of the model.</p></li>
</ul>
<p>Although the F-test is at the end, it should be examined first to determine whether the linear model fits the data if you have multiple response variables. Only when the model is well fitted, should you further interpret the significance of the intercept and slope.</p>
<p>In this example, the linear regression is significant because the p-value of the F test is lower than 0.05 (p = 0.000117), so we can continue to examine the intercept and slope.</p>
<p>The slope is 1.5556 (under “Estimate” in “Coefficient”) and the t-test indicates that it is significantly different from zero (p = 0.000117). Note that this p-value is the same as that of the F-test only when there is a single predictor variable. In multiple regression, there will be separate t-tests for each predictor variable.</p>
<p>Therefore we can write a statement to summarize the result and provide the statistical support:</p>
<blockquote class="blockquote">
<p><strong>There is a positive relationship between height and age (slope =</strong> 1.556<strong>, F = 48.47, d.f. = 1 and 8, p =</strong> 0.0001<strong>).</strong></p>
</blockquote>
<section id="visualization-scatter-plot" class="level3" data-number="9.1.1">
<h3 data-number="9.1.1" class="anchored" data-anchor-id="visualization-scatter-plot"><span class="header-section-number">9.1.1</span> Visualization: scatter plot</h3>
<p>Visualization for linear regression is shown above in <a href="#how-to-construct-a-scatter-plot-with-a-linear-fit">How to Construct a Scatter Plot with a Linear Fit</a>. For this dataframe <code>chirps</code>, we can do this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> chirps, <span class="fu">aes</span>(<span class="at">x =</span> temperature, <span class="at">y =</span> chirps_per_min)) <span class="sc">+</span> </span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> lm, <span class="at">se =</span> F, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Chirps per minute"</span>, <span class="at">x =</span> <span class="fu">expression</span>(<span class="st">"Temperature "</span> ( degree<span class="sc">*</span>C)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><strong>Figure 6. The relationship between air temperature and chirps per minute of snowy tree crickets.</strong> Chirps were counted for 10 crickets at different temperatures. There was a significant positive relationship between temperature and cricket’s chirps per minute (linear regression, p&lt;0.0001, R<sup>2</sup> = 0.85, y =-26.2+1.57*x).</p>
</section>
</section>
<section id="t-test" class="level2" data-number="9.2">
<h2 data-number="9.2" class="anchored" data-anchor-id="t-test"><span class="header-section-number">9.2</span> T-test</h2>
<p><strong>A t-test is used when the independent variable is nominal and has exactly two categories (such as an experimental group and a control group) and the dependent variable consists of continuous (measurement) data.</strong> If the p value is less than 0.05, then the statistical value you calculate (the ‘t-value’) has a low probability (less than 5%) of occurring by chance if the null hypothesis is true. What this means in practical terms is that you have evidence that something other than the null hypothesis is occurring - there is a relationship between your two variables! - and you can reject your null hypothesis and describe the relationship (in this case a relationship between the independent variable “experimental group” and the dependent variable “outcome measure” can also be described as a difference in outcome between groups). In contrast, if the p value is more than 0.05, then the biologist fails to reject the null hypothesis that there is no significant difference between the two groups. In practical terms this means that the scientist has no evidence that the means are different.</p>
<p>Let’s say that you want to test if a fertilizer has a significant effect on the growth of plants. You measure the height of 10 plants, 5 grown with just soil and 5 grown with soil and added fertilizer. Both groups are grown for the same length of time. Here are the instructions for performing a t-test (for comparing two groups) using R:</p>
<p>You can try the codes using the sample 210 data <a href="https://drive.google.com/drive/folders/12eJ87Si36TMe8bzovtc8L7mYVhi5_1HS?usp=drive_link">“2024 R Workshop for Faculty” Google Drive</a>.</p>
<p>Run this code to generate the data manually.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>fertilizer <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="st">"Treatment"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"Soil"</span>, <span class="dv">5</span>), <span class="fu">rep</span>(<span class="st">"Soil+fertilizer"</span>, <span class="dv">5</span>)),</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Plant_height_cm"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="fl">10.5</span>, <span class="dv">13</span>, <span class="dv">9</span>, <span class="dv">11</span>, <span class="dv">14</span>, <span class="dv">17</span>, <span class="fl">18.5</span>, <span class="dv">20</span>, <span class="dv">16</span>, <span class="fl">17.5</span>))</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>fertilizer</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Treatment Plant_height_cm
1             Soil            10.5
2             Soil            13.0
3             Soil             9.0
4             Soil            11.0
5             Soil            14.0
6  Soil+fertilizer            17.0
7  Soil+fertilizer            18.5
8  Soil+fertilizer            20.0
9  Soil+fertilizer            16.0
10 Soil+fertilizer            17.5</code></pre>
</div>
</div>
<p>Notice that there are only 2 nominal variables (soil and soil+fertilizer). If there are more than 2 nominal variables an ANOVA must be used to analyze the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(fertilizer<span class="sc">$</span>Treatment)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Soil"            "Soil+fertilizer"</code></pre>
</div>
</div>
<p>To test whether <code>Plant_height_cm</code> is different between <code>Treatment</code> , in other word, whether <code>Treatment</code> affect <code>Plant_height_cm</code>, we will use <code>t.test()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(Plant_height_cm<span class="sc">~</span>Treatment, <span class="at">data =</span> fertilizer, <span class="at">alternative =</span> <span class="st">"two.sided"</span>, <span class="at">var.equal =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  Plant_height_cm by Treatment
t = -5.6014, df = 7.4757, p-value = 0.0006475
alternative hypothesis: true difference in means between group Soil and group Soil+fertilizer is not equal to 0
95 percent confidence interval:
 -8.925619 -3.674381
sample estimates:
           mean in group Soil mean in group Soil+fertilizer 
                         11.5                          17.8 </code></pre>
</div>
</div>
<p><code>alternative</code> can also be “greater” or “less” if you have an a priori reason to test a directional difference between groups.</p>
<p>In this case, the p-value is 0.0006475. We failed to reject the null hypothesis that <code>Treatment</code> can affect <code>Plant_height_cm</code>. Hence we can write a summary statement like:</p>
<blockquote class="blockquote">
<p>Treatment group does not significantly affect plant height (t = -5.601, d.f. = 7.476, p = 0.001)</p>
</blockquote>
<section id="visualization-boxplot" class="level3" data-number="9.2.1">
<h3 data-number="9.2.1" class="anchored" data-anchor-id="visualization-boxplot"><span class="header-section-number">9.2.1</span> <strong>Visualization: boxplot</strong></h3>
<p>You can plot the continuous variable from two groups in a number of ways.</p>
<p>One way is to use a boxplot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> fertilizer, <span class="fu">aes</span>(<span class="at">x =</span> Treatment, <span class="at">y =</span> Plant_height_cm)) <span class="sc">+</span> </span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">outlier.colour =</span> <span class="st">"grey"</span>) <span class="sc">+</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">width =</span> <span class="fl">0.1</span>, <span class="at">height =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Treatment"</span>, <span class="at">y =</span> <span class="st">"Plant height (cm)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="anova" class="level2" data-number="9.3">
<h2 data-number="9.3" class="anchored" data-anchor-id="anova"><span class="header-section-number">9.3</span> ANOVA</h2>
<section id="global-anova-test" class="level3" data-number="9.3.1">
<h3 data-number="9.3.1" class="anchored" data-anchor-id="global-anova-test"><span class="header-section-number">9.3.1</span> Global ANOVA test</h3>
<p><strong>An ANOVA is used in a similar circumstance as a t-test: nominal independent variable and continuous dependent variable, but when there are three or more categories for the independent variable (perhaps you tried multiple experimental treatments)</strong>.</p>
<p>If the p value is more than 0.05, then the biologist fails to reject the “null hypothesis” that there is no significant difference between any pairs of groups. If the p value is less than 0.05, then the statistical value you calculate (the ‘F-value’) has a low probability (less than 5%) of occurring by chance if the null hypothesis is true. In this case, the null hypothesis is rejected, and the biologist must then perform additional “post-hoc” tests to determine exactly which pair(s) of groups are significantly different from one another. In the JMP output, groups with the same letters (such as A and A) are not significantly different from one another, while groups with different letters (such as A and B) are significantly different from one another (refer to sample graph). In the case of a significant difference, biologists must determine if the outcome is what they expected or is the opposite of what they expected.&nbsp;</p>
<p>Let’s say you want to measure the effectiveness of penicillin in disrupting E. coli growth. Penicillin is a common antibiotic that disrupts cell wall formation thus halting cellular division in many microbes. In order to see which dose is most effective, we will test 3 different concentrations as well as a control (no penicillin). The response variable in this case is measured in CFU (colony forming unit). This colony shows that how many single cells survived to divide into a visible colony.</p>
<p>Run this code to generate the data manually.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>penicillin <span class="ot">=</span> <span class="fu">data.frame</span>(</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">Penicillin_conc =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">0</span>,<span class="dv">5</span>), <span class="fu">rep</span>(<span class="dv">100</span>, <span class="dv">5</span>), <span class="fu">rep</span>(<span class="dv">150</span>, <span class="dv">5</span>), <span class="fu">rep</span>(<span class="dv">200</span>,<span class="dv">5</span>)), </span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">CFU =</span> <span class="fu">c</span>(<span class="dv">795</span>, <span class="dv">611</span>, <span class="dv">703</span>, <span class="dv">698</span>, <span class="dv">675</span>, <span class="dv">652</span>, <span class="dv">696</span>, <span class="dv">702</span>, <span class="dv">689</span>, <span class="dv">709</span>, <span class="dv">196</span>, <span class="dv">75</span>, <span class="dv">111</span>, <span class="dv">77</span>, <span class="dv">154</span>, <span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">0</span>))</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a><span class="co"># We will adjust Penicillin_conc to be a ordered factor</span></span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>penicillin<span class="sc">$</span>Penicillin_conc <span class="ot">=</span> <span class="fu">factor</span>(penicillin<span class="sc">$</span>Penicillin_conc, <span class="at">ordered =</span> T, <span class="at">levels =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">100</span>, <span class="dv">150</span>, <span class="dv">200</span>))</span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>penicillin</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Penicillin_conc CFU
1                0 795
2                0 611
3                0 703
4                0 698
5                0 675
6              100 652
7              100 696
8              100 702
9              100 689
10             100 709
11             150 196
12             150  75
13             150 111
14             150  77
15             150 154
16             200   2
17             200   0
18             200   2
19             200   1
20             200   0</code></pre>
</div>
</div>
<p>To test whether <code>Penicillin_conc</code> affects <code>CFU</code>, we will build the model using <code>aov()</code>, then use <code>summary()</code> to show the full results:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>p_anova_model <span class="ot">=</span> <span class="fu">aov</span>(<span class="at">formula =</span> CFU <span class="sc">~</span> Penicillin_conc, <span class="at">data =</span> penicillin)</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(p_anova_model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                Df  Sum Sq Mean Sq F value  Pr(&gt;F)    
Penicillin_conc  3 2029149  676383   356.4 7.4e-15 ***
Residuals       16   30362    1898                    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>Note that the results is similar to the F-test in the last part of linear regression, because ANOVA and regression are both linear models. But in ANOVA, we don’t have predictions of slope and intercepts.</p>
<p>In this case, since the p-value is &lt;0.05. We rejected the null hypothesis that <code>Penicillin_conc</code> has an effect on <code>CFU</code> Therefore, we can write a summary statement to say:</p>
<blockquote class="blockquote">
<p>Penicillin concentration significantly affect CFU (F = 356.4, d.f. = 3 and 16, p &lt; 0.0001)</p>
</blockquote>
</section>
<section id="post-hoc-test" class="level3" data-number="9.3.2">
<h3 data-number="9.3.2" class="anchored" data-anchor-id="post-hoc-test"><span class="header-section-number">9.3.2</span> <strong><em>Post hoc test</em></strong></h3>
<p><u><em>If the p value is below 0.05, you must run post hoc tests to compare the means.</em></u> <em>We can compare all means using</em> <strong>Tukey HSD</strong> <em>to obtain information about which pair(s) of species are significantly different from one another.</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(multcomp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: mvtnorm</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: survival</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: TH.data</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: MASS</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'TH.data'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:MASS':

    geyser</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Tukey HSD test:</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>post_test <span class="ot">&lt;-</span> <span class="fu">glht</span>(p_anova_model,</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">linfct =</span> <span class="fu">mcp</span>(<span class="at">Penicillin_conc =</span> <span class="st">"Tukey"</span>)</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(post_test)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
     Simultaneous Tests for General Linear Hypotheses

Multiple Comparisons of Means: Tukey Contrasts


Fit: aov(formula = CFU ~ Penicillin_conc, data = penicillin)

Linear Hypotheses:
               Estimate Std. Error t value Pr(&gt;|t|)    
100 - 0 == 0      -6.80      27.55  -0.247  0.99450    
150 - 0 == 0    -573.80      27.55 -20.827  &lt; 0.001 ***
200 - 0 == 0    -695.40      27.55 -25.241  &lt; 0.001 ***
150 - 100 == 0  -567.00      27.55 -20.580  &lt; 0.001 ***
200 - 100 == 0  -688.60      27.55 -24.994  &lt; 0.001 ***
200 - 150 == 0  -121.60      27.55  -4.414  0.00208 ** 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
(Adjusted p values reported -- single-step method)</code></pre>
</div>
</div>
<p><em>This test compares the mean of each group to determine which are significantly different from each other. <code>Pr(&gt;|t|)</code> indicates the p values for each pair. This shows us that all the means of E. coli growth are different except for the 0 and 100 ug/mL concentrations. These differences can be indicated with letters on your final figure.</em></p>
<p>To generate the letter groupings, we need to run the same test with a different library:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(agricolae)</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>hsd <span class="ot">=</span> <span class="fu">HSD.test</span>(p_anova_model, <span class="at">trt =</span> <span class="st">"Penicillin_conc"</span>, <span class="at">group =</span> T)</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>hsd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$statistics
  MSerror Df  Mean       CV      MSD
   1897.6 16 377.4 11.54251 78.82304

$parameters
   test          name.t ntr StudentizedRange alpha
  Tukey Penicillin_conc   4         4.046093  0.05

$means
      CFU      std r       se Min Max Q25 Q50 Q75
0   696.4 66.18006 5 19.48127 611 795 675 698 703
100 689.6 22.27779 5 19.48127 652 709 689 696 702
150 122.6 52.08935 5 19.48127  75 196  77 111 154
200   1.0  1.00000 5 19.48127   0   2   0   1   2

$comparison
NULL

$groups
      CFU groups
0   696.4      a
100 689.6      a
150 122.6      b
200   1.0      c

attr(,"class")
[1] "group"</code></pre>
</div>
</div>
<p>Under $groups, the letters indicates the CFU at 0 and 100 ug/mL of panicillin are not different from each other as they are both in group “a”. But 150 ug/ml and 200 ug/ml are both significantly different from both 0 and 100 ug/mL as they are in different letter groups b and c, respectively.</p>
</section>
<section id="visualization-box-plot-and-bar-plot" class="level3" data-number="9.3.3">
<h3 data-number="9.3.3" class="anchored" data-anchor-id="visualization-box-plot-and-bar-plot"><span class="header-section-number">9.3.3</span> <strong>Visualization: box plot and bar plot</strong></h3>
<p>Similar to the t-test, we can plot the data used for ANOVA using a boxplot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> penicillin, <span class="fu">aes</span>(<span class="at">x =</span> Penicillin_conc, <span class="at">y =</span> CFU)) <span class="sc">+</span> </span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">width =</span> <span class="fl">0.1</span>, <span class="at">height =</span> <span class="dv">0</span>) <span class="sc">+</span> </span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">expression</span>(<span class="st">"Amount of penicillin "</span> (mu<span class="sc">*</span>g<span class="sc">/</span>mL)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can also incorporate results from the post hoc test to the figure:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We use the letters from hsd to build a dataframe to help us label the plot</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>temp_label <span class="ot">=</span> <span class="fu">merge</span>(hsd<span class="sc">$</span>means, hsd<span class="sc">$</span>groups, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"row.names"</span>, <span class="fu">names</span>(hsd<span class="sc">$</span>means)[<span class="dv">1</span>]))</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(temp_label)[<span class="dv">1</span>] <span class="ot">=</span> hsd<span class="sc">$</span>parameters<span class="sc">$</span>name.t</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>temp_label</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Penicillin_conc   CFU      std r       se Min Max Q25 Q50 Q75 groups
1               0 696.4 66.18006 5 19.48127 611 795 675 698 703      a
2             100 689.6 22.27779 5 19.48127 652 709 689 696 702      a
3             150 122.6 52.08935 5 19.48127  75 196  77 111 154      b
4             200   1.0  1.00000 5 19.48127   0   2   0   1   2      c</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Then we can add labels to the ggplot</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> penicillin, <span class="fu">aes</span>(<span class="at">x =</span> Penicillin_conc, <span class="at">y =</span> CFU)) <span class="sc">+</span> </span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">width =</span> <span class="fl">0.1</span>, <span class="at">height =</span> <span class="dv">0</span>) <span class="sc">+</span> </span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">expression</span>(<span class="st">"Amount of penicillin "</span> (mu<span class="sc">*</span>g<span class="sc">/</span>mL))) <span class="sc">+</span> </span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data =</span> temp_label, <span class="fu">aes</span>(<span class="at">x =</span> Penicillin_conc, </span>
<span id="cb74-7"><a href="#cb74-7" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">y =</span> Max<span class="sc">+</span><span class="fu">max</span>(temp_label<span class="sc">$</span>Max)<span class="sc">*</span><span class="fl">0.1</span>, </span>
<span id="cb74-8"><a href="#cb74-8" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">label =</span> groups))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Use of `temp_label$Max` is discouraged.
ℹ Use `Max` instead.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Similar to <a href="#how-to-construct-a-bar-graph">How to Construct a Bar Graph</a>, we can create a bar plot with error bars and the post hoc letter groupings.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>penicillin_summary <span class="ot">=</span> <span class="fu">summarySE</span>(<span class="at">data =</span> penicillin, <span class="at">measurevar =</span> <span class="st">"CFU"</span>, <span class="at">groupvars =</span> <span class="st">"Penicillin_conc"</span>)</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> penicillin_summary, <span class="fu">aes</span>(<span class="at">x =</span> Penicillin_conc, <span class="at">y =</span> CFU)) <span class="sc">+</span></span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>) <span class="sc">+</span> </span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin=</span>CFU<span class="sc">-</span>sd, <span class="at">ymax=</span>CFU<span class="sc">+</span>sd), <span class="at">width=</span>.<span class="dv">2</span>,</span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a>                 <span class="at">position=</span><span class="fu">position_dodge</span>(.<span class="dv">9</span>)) <span class="sc">+</span></span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">expression</span>(<span class="st">"Amount of penicillin "</span> (mu<span class="sc">*</span>g<span class="sc">/</span>mL))) <span class="sc">+</span> </span>
<span id="cb76-8"><a href="#cb76-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">data =</span> penicillin, <span class="fu">aes</span>(<span class="at">x =</span> Penicillin_conc, <span class="at">y =</span> CFU), <span class="at">width =</span> <span class="fl">0.2</span>, <span class="at">height =</span> <span class="dv">0</span>) <span class="sc">+</span> </span>
<span id="cb76-9"><a href="#cb76-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text</span>(<span class="at">data =</span> temp_label, <span class="fu">aes</span>(<span class="at">x =</span> Penicillin_conc, </span>
<span id="cb76-10"><a href="#cb76-10" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">y =</span> Max<span class="sc">+</span><span class="fu">max</span>(temp_label<span class="sc">$</span>Max)<span class="sc">*</span><span class="fl">0.1</span>, </span>
<span id="cb76-11"><a href="#cb76-11" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">label =</span> groups))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Use of `temp_label$Max` is discouraged.
ℹ Use `Max` instead.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-40-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="two-types-of-chi-square-x2-tests" class="level2" data-number="9.4">
<h2 data-number="9.4" class="anchored" data-anchor-id="two-types-of-chi-square-x2-tests"><span class="header-section-number">9.4</span> Two types of Chi-square (X<sup>2</sup>) tests</h2>
<p><strong>Chi-Square tests are used when both the dependent and independent variables are nominal.</strong></p>
<section id="goodness-of-fit-test" class="level4" data-number="9.4.0.1">
<h4 data-number="9.4.0.1" class="anchored" data-anchor-id="goodness-of-fit-test"><span class="header-section-number">9.4.0.1</span> Goodness of Fit test</h4>
<p>The <strong>Goodness of Fit test</strong> is used to determine if the observed result (i.e.&nbsp;collected during an experiment) is significantly different than the “expected” result (i.e.&nbsp;that based on the laws of probability). It involves calculating the degrees of freedom (DF), which corresponds to the following equation: DF = (C – 1) where C is the number of categories. It also involves calculating the Chi-square value (X<sup>2</sup>), which corresponds to the following equation: X<sup>2</sup> = S (D<sup>2</sup>/E) where D is the difference between the observed result (O) and expected result (E). The biologist determines if the X<sup>2</sup> value is less than the “critical” X<sup>2</sup> value for an experiment with a particular DF to determine if there is a significant difference (p value &lt; 0.05) between the observed and expected result.</p>
<p><img src="Images/chisq.png" id="chisq" class="img-fluid">.</p>
<p>Let’s say that you have crustaceans and test their preference for a light or dark environment. In this example there is only ONE variable: the observed preference (are they in the light or dark when you check on them?). You want to see if the distribution of crustaceans is random (roughly 50% in the light and 50% in the dark because they have no preference and distribute according to the laws of probability) OR if the distribution is non-random because they prefer either dark or light.</p>
<p>The <strong>null hypothesis</strong> for this test is that the variable of interest (light/dark environment) will have no impact on the distribution of the data. In other words, the null hypothesis assumes a random distribution, which means <strong>observed is no different from expected</strong>. If p&gt;0.05, you accept the null.</p>
<p>A <strong>significant result</strong> (p&lt;0.05) means the distribution of your <strong>observed data is different than expected</strong>; it is non-random.</p>
</section>
<section id="chi-square---goodness-of-fit-test" class="level3" data-number="9.4.1">
<h3 data-number="9.4.1" class="anchored" data-anchor-id="chi-square---goodness-of-fit-test"><span class="header-section-number">9.4.1</span> Chi-square - Goodness of Fit test</h3>
<p>The goodness of fit Chi-square test is used to test whether the distribution of a categorical variable is significantly different than an “expected” distribution. There is no predictor in this case.</p>
<blockquote class="blockquote">
<p>For example, in a species of crustacena, we counted the number of individuals that prefer a light and a dark treatment. We may expect the light:dark ratio to be 1:1 and we can test it with the chi-square goodness of fit test.</p>
</blockquote>
<p>We will create a file that may be similar to how your raw data will look like in a “long form”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>crust_sp1 <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">Crustacean =</span> <span class="st">"Species 1"</span>, <span class="at">Result =</span> <span class="fu">c</span>(<span class="st">"light"</span>, <span class="st">"light"</span>, <span class="st">"dark"</span>, <span class="st">"light"</span>, <span class="st">"light"</span>, <span class="st">"light"</span>, <span class="st">"light"</span>, <span class="st">"light"</span>, <span class="st">"dark"</span>, <span class="st">"dark"</span>, <span class="st">"light"</span>, <span class="st">"light"</span>, <span class="st">"light"</span>, <span class="st">"light"</span>, <span class="st">"light"</span>, <span class="st">"light"</span>, <span class="st">"light"</span>, <span class="st">"dark"</span>, <span class="st">"dark"</span>, <span class="st">"dark"</span>, <span class="st">"light"</span>, <span class="st">"light"</span>, <span class="st">"light"</span>, <span class="st">"light"</span>))</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>crust_sp1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Crustacean Result
1   Species 1  light
2   Species 1  light
3   Species 1   dark
4   Species 1  light
5   Species 1  light
6   Species 1  light
7   Species 1  light
8   Species 1  light
9   Species 1   dark
10  Species 1   dark
11  Species 1  light
12  Species 1  light
13  Species 1  light
14  Species 1  light
15  Species 1  light
16  Species 1  light
17  Species 1  light
18  Species 1   dark
19  Species 1   dark
20  Species 1   dark
21  Species 1  light
22  Species 1  light
23  Species 1  light
24  Species 1  light</code></pre>
</div>
</div>
<p>We can convert this into the “short form” :</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>crust_table <span class="ot">=</span> <span class="fu">table</span>(crust_sp1[,<span class="st">"Result"</span>])</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>crust_table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 dark light 
    6    18 </code></pre>
</div>
</div>
<p>We can then test whether the observed distribution of <code>Result</code> is equally distributed among the two groups (i.e., 1:1) using <code>chisq.test()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="fu">chisq.test</span>(<span class="at">x =</span> crust_table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Chi-squared test for given probabilities

data:  crust_table
X-squared = 6, df = 1, p-value = 0.01431</code></pre>
</div>
</div>
<p>The X-squared value (6) is the “Pearson” test chi square value. It is used, with the degrees of freedom (df) value, to obtain the p-value from a chi-squared table.</p>
<p>The null hypothesis of the test is that the groups are equally distributed. The p-value is &lt; 0.05, which means that we can reject the null hypothesis. Therefore, our observed distribution of Age_group is significantly different from 1:1.</p>
<p>If there is an a priori reason that the distribution should be, for example, 3:1, we can modify the code to test this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="fu">chisq.test</span>(<span class="at">x =</span> crust_table, <span class="at">p =</span> <span class="fu">c</span>(<span class="dv">3</span><span class="sc">/</span><span class="dv">4</span>,<span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Chi-squared test for given probabilities

data:  crust_table
X-squared = 32, df = 1, p-value = 1.542e-08</code></pre>
</div>
</div>
<p>In this second test, the p-value is &lt; 0.05, so we can reject the null hypothesis. Hence the distribution is significantly different from 3:1.</p>
<p><strong>Visualization</strong></p>
<p>There are several ways of visually representing X<sup>2</sup> results in either a figure or a table. Ask your instructor about their preference.</p>
<p><em>For a table</em>:</p>
<p>The table should be <u>a simplified version</u> of the table that JMP provided.&nbsp; Rather than trying to modify the JMP table, just start from scratch in Excel or Word and make your own. Adjust the sizes of the cells and other formatting as needed to make it easy to read. See Table 2 at the very end of the Chi-square section for a good example.*</p>
<p>Just as a graph has a figure legend, a table must be accompanied by a table heading.&nbsp; Similar to a figure legend, the table heading starts with a number and a title in bold, then gives a description of the result <u>and statistics</u> in complete sentences.&nbsp; One key difference, while figure legends should be placed directly below a figure, the table heading goes above the table.&nbsp; Also, if you have both figures (graphs or pictures) and tables in your paper, you number them separately. Again, see Table 2 in the next section for good example.*</p>
<p><img src="Images/chisq_table.png" id="fig-chisq_table" class="img-fluid"></p>
<p><em>For a graph</em>:</p>
<p>The type of graph used to display frequency (count) data is called a histogram.</p>
<p>You can plot the data for a goodness of fit Chi-square test by using the raw data in <code>crust_sp1</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> crust_sp1, <span class="fu">aes</span>(<span class="at">x =</span> Result)) <span class="sc">+</span> </span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Result"</span>, <span class="at">y =</span> <span class="st">"Count"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Alternative, you could also plot using the count data <code>crust_table</code> after converting it to a dataframe:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">data.frame</span>(crust_table), <span class="fu">aes</span>(Var1, Freq, <span class="at">fill =</span> Var1)) <span class="sc">+</span> </span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>() <span class="sc">+</span> </span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Result"</span>, <span class="at">y =</span> <span class="st">"Count"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-46-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fill tells R to color the bar by Var1</span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a><span class="co"># you can remove the side legned by adding:</span></span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a><span class="co"># + theme(legend.position = "none")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="chi-square---contingency-test" class="level3" data-number="9.4.2">
<h3 data-number="9.4.2" class="anchored" data-anchor-id="chi-square---contingency-test"><span class="header-section-number">9.4.2</span> Chi-square - Contingency test</h3>
<p>The <strong>Contingency test,</strong> also called the <strong>Test of Independence</strong>, is used to determine if the difference between two groups is significantly different depending on <strong>(contingent on)</strong> another variable.&nbsp; Or, put another way, you might want to determine if the frequency of outcomes for one set of observations matches those for another set of observations.&nbsp; For example, let’s imagine that we want to know whether two different species of crustaceans exhibit the same preference for light or dark.&nbsp; In other words, is the preference for light or dark contingent on which species it is?&nbsp; To find this out one uses a special version of Chi-square tests known as a <em>contingency table</em>.</p>
<p>In this case, your <strong>independent variable</strong> is species and your <strong>dependent variable</strong> is preference (because you think preference might depend on species); your null hypothesis is that there is no difference in the preference for light versus dark between Species 1 and 2, regardless of whatever that preference is. When you run a contingency test you produce a X<sup>2</sup> value as in the Goodness of Fit test.&nbsp; If the X<sup>2</sup> value is large enough (leading to p&lt;0.05), then you can reject the null hypothesis that the proportion of crustaceans showing a preference for light is not different between the two species, and state that the data show that the preference for light or dark is contingent on the species.&nbsp; Another way to say the same thing: the two species show significantly different preferences for light.</p>
<p>Let’s imagine that we test the preference of two crustacean species for light versus dark and compare the preferences.&nbsp; The values could be arranged in a table like the one below.&nbsp;</p>
<p><img src="Images/Contingency_data.png" id="fig-Contingency_data" class="img-fluid"></p>
<p>If the preferences of the two species were the same, then they should each have roughly the same frequencies as the pooled results, which show animals split between dark and light, with possibly a small preference for dark (28 dark and 20 light). Do species 1 and 2 appear to have similar preferences?&nbsp; <em>Note that the total number of observations/ occurrences <u>do not</u> have to be the same for both species.</em>&nbsp; A contingency test will correct for different sizes between samples. Let’s conduct the test and see.</p>
<p>We will begin by making up the data for species 2:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>crust_sp2 <span class="ot">=</span>  <span class="fu">data.frame</span>(<span class="at">Crustacean =</span> <span class="st">"Species 2"</span>, <span class="at">Result =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"dark"</span>, <span class="dv">22</span>), <span class="fu">rep</span>(<span class="st">"light"</span>, <span class="dv">2</span>)))</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>crust_sp2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Crustacean Result
1   Species 2   dark
2   Species 2   dark
3   Species 2   dark
4   Species 2   dark
5   Species 2   dark
6   Species 2   dark
7   Species 2   dark
8   Species 2   dark
9   Species 2   dark
10  Species 2   dark
11  Species 2   dark
12  Species 2   dark
13  Species 2   dark
14  Species 2   dark
15  Species 2   dark
16  Species 2   dark
17  Species 2   dark
18  Species 2   dark
19  Species 2   dark
20  Species 2   dark
21  Species 2   dark
22  Species 2   dark
23  Species 2  light
24  Species 2  light</code></pre>
</div>
</div>
<p>Then we will add it to the previous <code>crust_sp1</code> to make a new object;</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>crust_twoSp <span class="ot">=</span> <span class="fu">rbind</span>(crust_sp1, crust_sp2)</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>crust_twoSp</span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a><span class="co"># The table is very long; so we are not displaying it here.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will first reshape our data into a contingency table using <code>table()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>crust_twoSp_table <span class="ot">=</span> <span class="fu">table</span>(crust_twoSp<span class="sc">$</span>Crustacean, crust_twoSp<span class="sc">$</span>Result)</span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>crust_twoSp_table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           
            dark light
  Species 1    6    18
  Species 2   22     2</code></pre>
</div>
</div>
<p>We can then test the null hypothesis that there is no difference in the preference for light versus dark between Species 1 and 2.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="fu">chisq.test</span>(crust_twoSp_table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Pearson's Chi-squared test with Yates' continuity correction

data:  crust_twoSp_table
X-squared = 19.286, df = 1, p-value = 1.125e-05</code></pre>
</div>
</div>
<p><strong>Visualization</strong></p>
<p>You can plot the contingency table by <code>geom_bar()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> crust_twoSp, <span class="fu">aes</span>(<span class="at">x =</span> Result)) <span class="sc">+</span> <span class="fu">geom_bar</span>() <span class="sc">+</span> <span class="fu">facet_grid</span>(<span class="sc">~</span>Crustacean) <span class="sc">+</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Result"</span>, <span class="at">y =</span> <span class="st">"Count"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The legend to this figure should be:</p>
<p><strong>Figure __. Two crustacean species showed different light/dark preferences.</strong> Twenty-four crustaceans of each species were observed to see if they moved to the light or dark portion of their tanks. The species were found to have significantly different preferences, with Species 1 being found in the light more often while Species 2 were more likely to be in the dark (Chi- square contingency test, Χ2 =24.442, d.f.= 1, p&lt;0.001).</p>
<p><em>Just as a graph has a figure legend, a table must be accompanied by a table heading.&nbsp; Similar to a figure legend, the table heading starts with a number and a title in bold, then gives a description of the result <u>and statistics</u> in complete sentences (Table 2).&nbsp; One key difference, while figure legends should be placed directly below a figure, the table heading goes above the table.&nbsp; Also, if you have both figures (graphs or pictures) and tables in your paper, you number them separately.</em></p>
<p><img src="Images/Contingency_table3.png" id="fig-Contingency_table3" class="img-fluid"> <img src="Images/Contingency_table4.png" class="img-fluid"></p>
</section>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>