<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2025-01-09">

<title>Big Red Statistics in R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="BigRedStatisticsInR_files/libs/clipboard/clipboard.min.js"></script>
<script src="BigRedStatisticsInR_files/libs/quarto-html/quarto.js"></script>
<script src="BigRedStatisticsInR_files/libs/quarto-html/popper.min.js"></script>
<script src="BigRedStatisticsInR_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="BigRedStatisticsInR_files/libs/quarto-html/anchor.min.js"></script>
<link href="BigRedStatisticsInR_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="BigRedStatisticsInR_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="BigRedStatisticsInR_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="BigRedStatisticsInR_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="BigRedStatisticsInR_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#i.-overview-of-the-scientific-method" id="toc-i.-overview-of-the-scientific-method" class="nav-link active" data-scroll-target="#i.-overview-of-the-scientific-method">I. Overview of the scientific method</a></li>
  <li><a href="#ii.-the-importance-of-statistical-tests" id="toc-ii.-the-importance-of-statistical-tests" class="nav-link" data-scroll-target="#ii.-the-importance-of-statistical-tests">II. The importance of statistical tests</a></li>
  <li><a href="#iii.-descriptive-statistics" id="toc-iii.-descriptive-statistics" class="nav-link" data-scroll-target="#iii.-descriptive-statistics">III. Descriptive statistics</a></li>
  <li><a href="#iv.-key-steps-in-data-analysis-using-r" id="toc-iv.-key-steps-in-data-analysis-using-r" class="nav-link" data-scroll-target="#iv.-key-steps-in-data-analysis-using-r">IV. Key steps in data analysis using R</a></li>
  <li><a href="#v.-start-using-r" id="toc-v.-start-using-r" class="nav-link" data-scroll-target="#v.-start-using-r">V. Start using R</a>
  <ul>
  <li><a href="#a.-access-r-via-posit-cloud-or-rstudio" id="toc-a.-access-r-via-posit-cloud-or-rstudio" class="nav-link" data-scroll-target="#a.-access-r-via-posit-cloud-or-rstudio">A. Access R via Posit Cloud or RStudio</a></li>
  <li><a href="#b.-project-setup-and-file-transfer" id="toc-b.-project-setup-and-file-transfer" class="nav-link" data-scroll-target="#b.-project-setup-and-file-transfer">B. Project setup and file transfer</a>
  <ul class="collapse">
  <li><a href="#online---posit-cloud" id="toc-online---posit-cloud" class="nav-link" data-scroll-target="#online---posit-cloud"><strong>Online - Posit Cloud</strong></a></li>
  <li><a href="#local---rstudio" id="toc-local---rstudio" class="nav-link" data-scroll-target="#local---rstudio"><strong>Local - RStudio</strong></a></li>
  </ul></li>
  <li><a href="#c.-starting-an-r-script" id="toc-c.-starting-an-r-script" class="nav-link" data-scroll-target="#c.-starting-an-r-script">C. Starting an R script</a></li>
  <li><a href="#d.-the-rstudio-interface" id="toc-d.-the-rstudio-interface" class="nav-link" data-scroll-target="#d.-the-rstudio-interface">D. The RStudio Interface</a></li>
  <li><a href="#e.-r-packages" id="toc-e.-r-packages" class="nav-link" data-scroll-target="#e.-r-packages">E. R packages</a>
  <ul class="collapse">
  <li><a href="#installing-packages" id="toc-installing-packages" class="nav-link" data-scroll-target="#installing-packages">Installing packages</a></li>
  <li><a href="#loading-packages" id="toc-loading-packages" class="nav-link" data-scroll-target="#loading-packages">Loading packages</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#vi.-importing-data-into-r" id="toc-vi.-importing-data-into-r" class="nav-link" data-scroll-target="#vi.-importing-data-into-r">VI. Importing data into R</a>
  <ul>
  <li><a href="#a.-preparing-a-.csv-file-from-excel" id="toc-a.-preparing-a-.csv-file-from-excel" class="nav-link" data-scroll-target="#a.-preparing-a-.csv-file-from-excel">A. Preparing a .csv file from Excel</a></li>
  <li><a href="#b.-loading-a-.csv-file-into-r" id="toc-b.-loading-a-.csv-file-into-r" class="nav-link" data-scroll-target="#b.-loading-a-.csv-file-into-r">B. Loading a .csv file into R</a></li>
  </ul></li>
  <li><a href="#vii.-how-to-generate-descriptive-statistics" id="toc-vii.-how-to-generate-descriptive-statistics" class="nav-link" data-scroll-target="#vii.-how-to-generate-descriptive-statistics">VII. How to generate descriptive statistics</a></li>
  <li><a href="#viii.-inferential-statistics-graphs" id="toc-viii.-inferential-statistics-graphs" class="nav-link" data-scroll-target="#viii.-inferential-statistics-graphs">VIII. Inferential statistics &amp; graphs</a>
  <ul>
  <li><a href="#a.-t-test" id="toc-a.-t-test" class="nav-link" data-scroll-target="#a.-t-test">A. T-test</a>
  <ul class="collapse">
  <li><a href="#two-sample-t-test" id="toc-two-sample-t-test" class="nav-link" data-scroll-target="#two-sample-t-test">Two-sample t-test</a></li>
  <li><a href="#visualization" id="toc-visualization" class="nav-link" data-scroll-target="#visualization"><strong>Visualization</strong></a>
  <ul class="collapse">
  <li><a href="#bar-graph" id="toc-bar-graph" class="nav-link" data-scroll-target="#bar-graph">Bar graph</a></li>
  <li><a href="#exporting-graphs" id="toc-exporting-graphs" class="nav-link" data-scroll-target="#exporting-graphs">Exporting graphs</a></li>
  <li><a href="#boxplot" id="toc-boxplot" class="nav-link" data-scroll-target="#boxplot">Boxplot</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#b.-anova" id="toc-b.-anova" class="nav-link" data-scroll-target="#b.-anova">B. ANOVA</a>
  <ul class="collapse">
  <li><a href="#global-anova-test" id="toc-global-anova-test" class="nav-link" data-scroll-target="#global-anova-test">Global ANOVA test</a></li>
  <li><a href="#post-hoc-test" id="toc-post-hoc-test" class="nav-link" data-scroll-target="#post-hoc-test"><strong><em>Post hoc test</em></strong></a></li>
  <li><a href="#visualization-1" id="toc-visualization-1" class="nav-link" data-scroll-target="#visualization-1"><strong>Visualization</strong></a>
  <ul class="collapse">
  <li><a href="#bar-graph-1" id="toc-bar-graph-1" class="nav-link" data-scroll-target="#bar-graph-1"><strong>Bar graph</strong></a></li>
  <li><a href="#box-plot" id="toc-box-plot" class="nav-link" data-scroll-target="#box-plot"><strong>Box plot</strong></a></li>
  </ul></li>
  </ul></li>
  <li><a href="#c.-linear-regression" id="toc-c.-linear-regression" class="nav-link" data-scroll-target="#c.-linear-regression">C. Linear regression</a>
  <ul class="collapse">
  <li><a href="#simple-linear-regression" id="toc-simple-linear-regression" class="nav-link" data-scroll-target="#simple-linear-regression">Simple linear regression</a></li>
  <li><a href="#visualization-2" id="toc-visualization-2" class="nav-link" data-scroll-target="#visualization-2">Visualization</a>
  <ul class="collapse">
  <li><a href="#scatter-plot-line-plot" id="toc-scatter-plot-line-plot" class="nav-link" data-scroll-target="#scatter-plot-line-plot">Scatter plot / Line plot</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#d.-two-types-of-chi-square-x2-tests" id="toc-d.-two-types-of-chi-square-x2-tests" class="nav-link" data-scroll-target="#d.-two-types-of-chi-square-x2-tests">D. Two types of Chi-square (X<sup>2</sup>) tests</a>
  <ul class="collapse">
  <li><a href="#chi-square---goodness-of-fit-test" id="toc-chi-square---goodness-of-fit-test" class="nav-link" data-scroll-target="#chi-square---goodness-of-fit-test"><strong>Chi-square - Goodness of Fit test</strong></a>
  <ul class="collapse">
  <li><a href="#visualization-3" id="toc-visualization-3" class="nav-link" data-scroll-target="#visualization-3"><strong>Visualization</strong></a></li>
  </ul></li>
  <li><a href="#chi-square---contingency-test" id="toc-chi-square---contingency-test" class="nav-link" data-scroll-target="#chi-square---contingency-test">Chi-square - Contingency test</a>
  <ul class="collapse">
  <li><a href="#visualization-4" id="toc-visualization-4" class="nav-link" data-scroll-target="#visualization-4"><strong>Visualization</strong></a></li>
  </ul></li>
  </ul></li>
  </ul></li>
  <li><a href="#appendix" id="toc-appendix" class="nav-link" data-scroll-target="#appendix">Appendix</a>
  <ul>
  <li><a href="#appendix-1.-editing-data-in-r" id="toc-appendix-1.-editing-data-in-r" class="nav-link" data-scroll-target="#appendix-1.-editing-data-in-r">Appendix 1. Editing data in R</a>
  <ul class="collapse">
  <li><a href="#changing-column-names" id="toc-changing-column-names" class="nav-link" data-scroll-target="#changing-column-names">Changing column names</a></li>
  <li><a href="#changing-values-in-a-data-frame" id="toc-changing-values-in-a-data-frame" class="nav-link" data-scroll-target="#changing-values-in-a-data-frame">Changing values in a data frame</a></li>
  <li><a href="#combining-text-from-two-columns" id="toc-combining-text-from-two-columns" class="nav-link" data-scroll-target="#combining-text-from-two-columns">Combining text from two columns</a></li>
  </ul></li>
  <li><a href="#appendix-2.-advanced-graphing" id="toc-appendix-2.-advanced-graphing" class="nav-link" data-scroll-target="#appendix-2.-advanced-graphing">Appendix 2. Advanced graphing</a>
  <ul class="collapse">
  <li><a href="#overlaying-data-points-in-bar-graph-and-boxplot" id="toc-overlaying-data-points-in-bar-graph-and-boxplot" class="nav-link" data-scroll-target="#overlaying-data-points-in-bar-graph-and-boxplot">Overlaying data points in bar graph and boxplot</a></li>
  <li><a href="#adding-post-hoc-grouping-into-bar-graph-and-boxplot" id="toc-adding-post-hoc-grouping-into-bar-graph-and-boxplot" class="nav-link" data-scroll-target="#adding-post-hoc-grouping-into-bar-graph-and-boxplot">Adding post hoc grouping into bar graph and boxplot</a></li>
  </ul></li>
  <li><a href="#appendix-3.-create-dataframes-in-r" id="toc-appendix-3.-create-dataframes-in-r" class="nav-link" data-scroll-target="#appendix-3.-create-dataframes-in-r">Appendix 3. Create dataframes in R</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Big Red Statistics in R</h1>
</div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 9, 2025</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="i.-overview-of-the-scientific-method" class="level1">
<h1>I. Overview of the scientific method</h1>
<p>Biologists are interested in organisms (either unicellular or multicellular), populations, and/or communities in a variety of contexts, including their molecular biology, physiology, growth, development, behavior, ecology, and evolution. Regardless of their specific interest, biologists usually rely on the <strong>scientific method</strong> to obtain information if it is not already available in the scientific literature. The scientific method consists of several steps: 1) making observations, 2) asking a question, 3) forming a hypothesis, 4) designing and executing an experiment to test the hypothesis, 5) analyzing the data (including a statistical test, if appropriate), 6) interpreting the results, and 7) communicating the results (usually with a table or graph).</p>
<p>An experiment often involves one or more <strong>variables</strong>; these may be <strong>numeric</strong> variables, expressed as numbers (e.g.&nbsp;“2.4 cm”), <strong>nominal</strong> variables, expressed as named categories (e.g.&nbsp;“female” vs.&nbsp;“male”), or <strong>ordinal</strong> variables, expressed as positions (e.g.&nbsp;“first’’ vs.&nbsp;``second” vs.&nbsp;“third”). The type of variable(s) determines the type of analysis that should be applied. In the case of numeric variables, the data may be <strong>continuous</strong> corresponding to an infinite range of values (e.g.&nbsp;as when measuring the length of a leaf) or <strong>discrete</strong>, corresponding to whole numbers (e.g.&nbsp;as when counting the number of stomata in a leaf).</p>
<p>In most cases, an experiment involves determining if a <strong>dependent variable</strong> is affected by an <strong>independent variable</strong> (e.g.&nbsp;as when counting the number of offspring – dependent - to determine if temperature – independent - has any effect on reproduction), or determining if there is a relationship between the two variables that may or may not involve cause and effect (e.g.&nbsp;as when you measure the weight of an organism to determine if it is related to the height of the organism).</p>
<p>The word “<strong>hypothesis</strong>” is used in slightly different ways in experimental design and statistics. Typically, when designing an experiment, the hypothesis is an educated guess about what you think might happen. It is generally expressed as a relationship between your two variables; e.g., plants that receive more light (independent variable) will grow taller (dependent variable). But when you go to analyze your data with statistics, you will use the term hypothesis with more nuance. Statistical tests help you determine if there is a significant relationship between your variables. When you do a statistical test, the default assumption is that there is NO relationship between your variables (amount of light has no impact on plant height). This default is called the “<strong>null hypothesis</strong>”. Thus, the null hypothesis is typically the opposite of the experimental hypothesis. The experimental hypothesis (referred to statistically as the alternative hypothesis) predicts a specific relationship between variables; <strong>the null hypothesis predicts no meaningful relationship between variables.</strong></p>
<p>If the results of an experiment are consistent with the null hypothesis, and thus do not support the biologist’s experimental hypothesis, then the researcher must conclude that there is no evidence of a significant relationship between the variables. The experimental hypothesis is either rejected or it is modified and subjected to additional testing. If the results of an experiment are inconsistent with the null hypothesis, then there IS a significant relationship between the variables which may support the experimental hypothesis, but the hypothesis may be subjected to additional testing as it is replicated and refined.</p>
<p>An experiment should be <strong>controlled</strong> to ensure that the results are not attributed to <strong>confounding variables</strong> (as when the dependent variable is affected by something other than the independent variable). In some cases, the biologist might utilize an <strong>experimental group</strong> (e.g.&nbsp;individuals that receive a treatment) as well as a <strong>control group</strong> (e.g.&nbsp;individuals that do not receive a treatment). In other cases, the biologist might utilize a <strong>positive control</strong> (which should produce results) and/or a <strong>negative control</strong> (which should not produce results). In all cases, it is ideal to perform multiple, identical trials (or replicates) and to test multiple subjects (i.e.&nbsp;a large <strong>sample size or large number of replicates</strong>) to ensure that there is a consistent outcome for the experiment.</p>
<p>There are additional ways to deal with confounding variables; these are by <strong>randomization</strong> and by <strong>matching</strong>. If you randomize the collection of your data, then you avoid the possibility that the effects you observe are due to confounding variables. For example, if I’m testing the effects of fertilizer on growth, but I chose all of my control plants in an area full of sunlight and all of my experimental plants in an area of heavy shade, then the differences in growth due to my fertilizer manipulation could be masked by the differences in sunlight. The remedy in this case is to randomly select control and experimental plants in both shady and sunny areas. Another way to control for confounding variables is through <strong>matching</strong> your control and experimental samples. As in the example above, if I placed a control and an experimental plant near each other (‘matching the samples’) in each habitat then the differences I observe are less likely due to differences in environmental variables that could confound my results.</p>
<p>A hypothesis becomes a <strong>theory</strong> if the results of many additional, independent tests continue to support it. A theory is an explanation that is supported by such a wide body of evidence that it is accepted as fact. Thus, biologists’ use of the term “theory” differs significantly from that of the general public who use the term when referring to speculation. One example of a theory is the theory of evolution, which is accepted as fact although there is still much for biologists to learn about natural selection and other mechanisms that result in a change in the relative frequency of alleles within a population over different time scales (i.e.&nbsp;micro versus macroevolution).</p>
<hr>
</section>
<section id="ii.-the-importance-of-statistical-tests" class="level1">
<h1>II. The importance of statistical tests</h1>
<p>Biologists often apply a statistical test for a less biased approach to interpreting the results of an experiment. A statistical test will estimate the probability of producing the observed result under the null hypothesis. In particular, it will yield a <strong>p value</strong>, which is the probability of producing the observed result by chance when there is actually no relationship between variables (e.g., when the null hypothesis is true). The p value ranges from 0 to 1, with 0 being no chance of producing this data set by chance if there is no relationship, and 1 being a 100% chance of producing this data set when there is no relationship. <strong>If the p value is less than 0.05</strong>, then the null hypothesis is rejected and the results are referred to as “significant,” indicating that there is a significant relationship between the variables. It is important to note that the value is a probability and thus the conclusion you draw from the p value could still be inaccurate as in the case of <strong>false positives</strong> (referred to as Type I errors), in which the data deviate from the null hypothesis by chance, and <strong>false negatives</strong> (referred to as Type II errors), in which the data erroneously support the null hypothesis (often due to the sample size being limited). Let’s reiterate those definitions:</p>
<ul>
<li><p>Type I error (false positive): for your sample, the p value is &lt;0.05 indicating a relationship between variables, but there would be no relationship if you measured a different or larger population.</p></li>
<li><p>Type II error (false negative): for your sample, the p value is &gt;0.05 indicating no relationship between variables, but there would be a relationship if you measured a different or larger population.</p></li>
</ul>
<p>The threshold that marks significance for the p value is referred to as the <strong>α value</strong>. The choice of α = 0.05 is standard, but arbitrary. It is possible to select a different significance level prior to the experiment. However, if you increase your significance level, say to 0.1, you increase your chance of a false positive, and if you decrease your significance level, say to 0.01, you increase your chance of a false negative.</p>
<p><img src="Images/flow.png" class="img-fluid"></p>
<hr>
</section>
<section id="iii.-descriptive-statistics" class="level1">
<h1>III. Descriptive statistics</h1>
<p><strong><em>Mean</em></strong></p>
<p>The <strong>mean (</strong><span class="math inline">\(\bar{X}\)</span><strong>)</strong> (often referred to as the average) corresponds to the sum of the measurements divided by the number of measurements (as in the case of 10, 40, 25, 35, and 40, which produce a mean of 30). In the case of an odd number of measurements, the <strong>median</strong> corresponds to the measurement in the middle when all of the measurements are sorted from lowest to highest (as in the case of 10, 40, 25, 35, and 40, which produces a median of 35). In the case of an even number of measurements, the median corresponds to the mean of the two measurements in the middle when all of the measurements are sorted from lowest to highest. Finally, the <strong>mode</strong> corresponds to the most common measurement (as in the case of 10, 40, 25, 35, and 40, which produces a mode of 40).</p>
<p><strong><em>Standard deviation</em></strong></p>
<p>The standard deviation (s) is a measure of the distribution of the data (i.e.&nbsp;extent of variation among different subjects) and weights each value by its distance from the mean. There is an increase in the standard deviation if there is a large distance of the values from the mean (as in the case of 10, 40, 25, 35, 40, which have a mean of 30). In contrast, there is a decrease in the standard deviation if there is a small distance (if any) of the values from the mean (as in the case of 25, 30, 35, 30, 30, which have the same mean of 30). The result for each group is summarized as the mean “plus or minus” the standard deviation (<span class="math inline">\(\bar{X}\)</span> ± s). This gives you a lot of information about the data set. If someone describes a data set as the mean plus or minus one SD, they are giving you the range into which 68% of the data points fall. Mean ± 2 SD gives you the range of 95% of the data, and Mean ± 3 SD encompasses 99% of the data. Note that on a bar graph or boxplot the standard deviation is sometimes represented by vertical lines known as error bars that extend from the mean in both an upward and downward direction.</p>
<p><img src="Images/errorbar.png" class="img-fluid"></p>
<p><strong><em>Standard error</em></strong></p>
<p>The standard error of the mean (SEM), or simply the <strong>standard error</strong> (SE) is another common way to state the variation around the mean. Instead of telling you about the range of the data set, it tells you the range of the “true” mean. If you added one sample to your study or took one sample away, the mean may change a little. So your mean is correct for your group of samples, but not necessarily for the whole population of all organisms that could have been in your study. SEM tells you that if you could have included every possible subject or sample - the mean probably would have been somewhere in this range. Once you have calculated the standard deviation of the mean (s), it is easy to calculate SE.</p>
<p><img src="Images/SE.png" class="img-fluid" width="197"></p>
<p>n = the number of samples or trials or replicates. This technique results in a smaller number than the standard deviation (s) offers, and thus often makes graphs with error bars look smaller. Whether to use s or SE in reporting your statistics and making error bars varies depending on the application, style, and field. Be sure to find out what your instructor prefers for your assignments!</p>
<p><strong><em>Coefficient of variation</em></strong></p>
<p>The <strong>coefficient of variation (CV)</strong> corresponds to the standard deviation divided by the mean and is useful when comparing the amount of variation for one variable among groups with different means.</p>
<p><br>
The variance in data sometimes fits a <strong>normal distribution</strong> (i.e.&nbsp;bell curve on a graph); such data are subject to a parametric test. However, the variance in data sometimes does not fit a normal distribution; such data are subject to a non-parametric test (addressed in advanced coursework).</p>
<hr>
</section>
<section id="iv.-key-steps-in-data-analysis-using-r" class="level1">
<h1>IV. Key steps in data analysis using R</h1>
<p>Here is the basic steps to analyze your data in R.</p>
<ol type="1">
<li><p>Access Posit Cloud (<a href="https://posit.cloud/" class="uri">https://posit.cloud/</a>) online or RStudio on your computer.</p></li>
<li><p>Create or open a project.</p></li>
<li><p>Create or open an R script.</p></li>
<li><p>If you are starting a new project, install a set of required packages by entering a code in the Script Panel and running it.</p></li>
<li><p>Prepare R to use those packages by ‘loading’ the packages. Do this every time!</p></li>
<li><p>In Excel, save your data into a .csv file. If needed, edit your data in Excel.</p></li>
<li><p>Upload the .csv file into Posit Cloud or RStudio.</p></li>
<li><p>Read your data file into R by entering a code.</p></li>
<li><p>If needed, generate descriptive statistics on one or more of the variables in your data set.</p></li>
<li><p>If not done prior to data collection, decide which inferential statistical test to use, such as a t-test, ANOVA, linear regression, or Chi-squared test.</p></li>
<li><p><strong>Locate the general code for the test you want to use and type (or copy &amp; paste) the code into the Script Panel in R.</strong></p></li>
<li><p><strong>If needed, substitute the names of your data file and variables into the general code.</strong></p></li>
<li><p><strong>Press <code>Run</code> in the upper right corner of the script panel. The output will appear in the Console Panel.</strong></p></li>
<li><p>Trouble-shoot errors that show up in the Console Panel.</p></li>
<li><p>Once the code is running correctly, decide what the output is telling you. For instance, do the findings lead you to accept or reject the test’s null hypothesis?</p></li>
<li><p>If needed, run a post hoc test.</p></li>
<li><p>Visualize the findings of your statistical test by making a graph in R or table in Excel.</p></li>
<li><p>Export or copy your graph into your lab assignment or into your research paper. Also record the needed statistical output, such a F or P values, so you can finish your lab assignment or use the output when writing a research paper.</p></li>
<li><p>Save the R script.</p></li>
</ol>
<hr>
</section>
<section id="v.-start-using-r" class="level1">
<h1>V. Start using R</h1>
<section id="a.-access-r-via-posit-cloud-or-rstudio" class="level2">
<h2 class="anchored" data-anchor-id="a.-access-r-via-posit-cloud-or-rstudio">A. Access R via Posit Cloud or RStudio</h2>
<p>R is a command line program in which you interact through text (the command line) instead of a graphical user interface (e.g., Windows, Excel, etc.). We provide two options for running R through an integrative environment called RStudio. In both options, you will not interact directly with R. You will run RStudio through Posit Cloud or RStudio software downloaded to your computer. Your instructor will let you know how you should use R for the class.</p>
<ul>
<li><p>The online version of RStudio is called <strong>Posit Cloud</strong>. Sign up for an account <a href="https://posit.co/products/cloud/cloud/">here</a>. The “Cloud Free” option will be sufficient. Do not choose the PLUS or other options. Your instructor will send you a link to get access to a shared class space, which has more memory and run time allocations than the “Cloud free” account.</p></li>
<li><p>If directed by your instructor, you can download RStudio and install it on your computer. Depending on the version of RStudio and R, the interface and commands could be slightly different from the online version. Follow the instructions <a href="https://posit.co/download/rstudio-desktop/">here</a> to download and install both R and RStudio.</p></li>
</ul>
</section>
<section id="b.-project-setup-and-file-transfer" class="level2">
<h2 class="anchored" data-anchor-id="b.-project-setup-and-file-transfer">B. Project setup and file transfer</h2>
<section id="online---posit-cloud" class="level3">
<h3 class="anchored" data-anchor-id="online---posit-cloud"><strong>Online - Posit Cloud</strong></h3>
<p>For the online option, you will need to upload data to the cloud or download output files from the cloud. The benefit is that you can use R anywhere and on virtually any machine with the internet (even a tablet)!</p>
<p>Once in the classroom space, you will see project assignments that your instructor has created for you. Your instructor will also have access to your assignments in this space. In some cases, your instructor may have created a template that includes pre-installed packages.</p>
<p>To create your own project on Posit, click <code>New Project &gt; New RStudio Project</code> on the upper right. You should rename your project(s) as directed by your instructor.</p>
<p><img src="Images/NewProject.png" id="fig-NewProject" class="img-fluid" width="452"></p>
<p>Once the project is deployed, on the lower right panel, click the <code>Upload</code> button and choose the data file from your computer.</p>
<p><img src="Images/upload.png" id="fig-upload" class="img-fluid" width="462"></p>
<p>To download a file, check the box next to the file you want to export, click <code>More &gt; Export…</code> and click <code>Download</code>.</p>
<p><img src="Images/Export.png" id="fig-export" class="img-fluid" width="451"></p>
</section>
<section id="local---rstudio" class="level3">
<h3 class="anchored" data-anchor-id="local---rstudio"><strong>Local - RStudio</strong></h3>
<p>If you choose the local option, you should set up a project and a folder that is associated with this project where all files will be stored.</p>
<p>For example, create a folder on your desktop that is called “R_is_fun”. Start RStudio, click <code>File &gt; New Project…</code> and select <code>Existing Directory</code> and browse to the folder <code>R_is_fun</code>.</p>
<p>Then click <code>Create Project</code>. Now you can access all files in the folder <code>R_is_fun</code> (without indicating the full pathname). Everything you saved in R will also appear in this folder.</p>
</section>
</section>
<section id="c.-starting-an-r-script" class="level2">
<h2 class="anchored" data-anchor-id="c.-starting-an-r-script">C. Starting an R script</h2>
<p>Once your project is opened, you should create an R script. The R script is a text file in which your R commands and notes will be saved.</p>
<p>To do this, click <code>File &gt; New File &gt; R Script</code>.</p>
<p>To save the R script, click <code>File &gt; Save</code> and give it an informative name. You should rename your R script(s) as directed by your instructor. The R script will be saved to a file with a .R extension. It is a text file that can be opened in any text editor.</p>
<p>If you are opening an existing project, the R scripts from your previous run are typically opened automatically.</p>
</section>
<section id="d.-the-rstudio-interface" class="level2">
<h2 class="anchored" data-anchor-id="d.-the-rstudio-interface">D. The RStudio Interface</h2>
<p>Let’s introduce what is being displayed in Posit and RStudio. With an R script opened, you should see a default four-panel interface.</p>
<p><img src="Images/Rstudio_interface.png" id="fig-RStudio_interface" class="img-fluid"></p>
<ul>
<li><p><span class="math inline">\(A\)</span> The upper left is the R <strong>Script Panel</strong>. This is where you will write and save your R script or commands. The panel should be saved regularly, which serves as a record of all the commands that you have (or are planning) to run. The scripts here are not being read by R until you click the <code>Run</code> button on the top right.</p></li>
<li><p><span class="math inline">\(B\)</span> The lower left is the R <strong>Console Panel</strong>, which is the innate command line interface of R. A command from the R script above will be displayed here when you click the <code>Run</code> button, which is when R “reads” your command. Any output or error from R will also be displayed here. You could type your command directly here, but it will not be saved in the R script.</p></li>
<li><p><span class="math inline">\(C\)</span> The upper right is the <strong>Environment Panel</strong>. Objects that are read or generated in R will be displayed here.</p></li>
<li><p><span class="math inline">\(D\)</span> The lower right is the <strong>Files Panel</strong>. It displays the files in your working directory. In Posit cloud, you will see files uploaded to the course space by your instructor. It also displays figures in the <code>Plots</code> tab if your commands result in graphical outputs. The <code>Packages</code> tab indicates which packages are currently installed, while check marks indicate which ones are loaded in the library. The <code>Help</code> tab displays the manual for each package and command.</p></li>
</ul>
<p>For a more comprehensive introduction to RStudio, see <a href="https://education.rstudio.com/learn/beginner/">here</a>.</p>
</section>
<section id="e.-r-packages" class="level2">
<h2 class="anchored" data-anchor-id="e.-r-packages">E. R packages</h2>
<p>Many packages are available to expand the basic functions of base R. This is similar to adding a plug-in or extension to your web browser.</p>
<section id="installing-packages" class="level3">
<h3 class="anchored" data-anchor-id="installing-packages">Installing packages</h3>
<p>To use these packages, you have to install them when you start the project for the first time. You will not need to install them again when you start an old project. Then you can load the package as libraries.</p>
<p>We will install a few packages that will be used in this document:</p>
<ul>
<li><p>Copy or type lines of codes below and paste them into the Script Panel (top left, Panel A), exactly as shown. The gray box below is called a “code chunk”. In the online version of this guide, you can copy everything from a code chunk using the copy button that appears when you hover your cursor on the code chunk.</p>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="Images/copy_code_chunk.png" class="img-fluid figure-img"></p>
</figure>
</div></li>
<li><p>Highlight all the codes and hit the <code>Run</code> button on the top right of the Script Panel. Alternatively, use the keyboard shortcut <code>Ctrl+Enter</code> (PC) or <code>Cmd+Enter</code> (Mac).</p></li>
<li><p>You may be asked to select a mirror to download the packages and enter <code>Y</code> to update packages.</p></li>
<li><p>A number of output should show up in the Console Panel to report the installation progress.</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Run these only once when you first start the project:</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"Rmisc"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"reshape"</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"agricolae"</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"multcomp"</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="co"># In most command-line interfaces, words after # are not read by the software and serve as notes.</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="co"># We will explain the R syntax below in "Importing data into R"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Alternatively, you can install packages from the <code>Package</code> tab on the Files Panel (Panel D). Click <code>Install</code> and type the package name in the box (it will auto-complete for you to prevent typos).</p>
<p><img src="Images/install_package.png" class="img-fluid" width="322"></p>
</section>
<section id="loading-packages" class="level3">
<h3 class="anchored" data-anchor-id="loading-packages">Loading packages</h3>
<p>Then, we will load these packages such that functions from these packages can be used. <em>Note: You WILL need to run library() for each package EVERY TIME you start a new Project. But you will not need to install them again.</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Run these every time before you run anything else:</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"Rmisc"</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"reshape"</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"agricolae"</span>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"multcomp"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
</section>
<section id="vi.-importing-data-into-r" class="level1">
<h1>VI. Importing data into R</h1>
<section id="a.-preparing-a-.csv-file-from-excel" class="level2">
<h2 class="anchored" data-anchor-id="a.-preparing-a-.csv-file-from-excel">A. Preparing a .csv file from Excel</h2>
<p>Your data should be entered and stored in spreadsheet programs such as Microsoft Excel before being imported into R. While R can read Excel files (A) such as .xls and .xlsx, data are typically saved into a simpler text format such as .csv (comma-separated values, B), where each row appears in a new line and each column is separated by a comma (,). Note that there is a header line (the first row) that specifies the column names.</p>
<p><img src="Images/excel_vs_csv.png" id="fig-excel" class="img-fluid"></p>
<p>A spreadsheet in Excel can be exported into .csv by clicking <code>File &gt; Save As…</code> and selecting <code>CSV</code> in the drop-down manual under <code>File Format</code>.</p>
<p>Try doing this with <code>name.xlsx</code> which can be downloaded from <a href="https://github.com/solomonchak/R_Intro/raw/5d08edb7d4981d78066a6a5294567e735fd00350/File/name.xlsx">here</a> (<a href="https://tinyurl.com/2k2aksxf" class="uri">https://tinyurl.com/2k2aksxf</a>)</p>
<p>Place the .csv file into your project’s directory (e.g., R_is_fun) or upload it to your project in posit.cloud using the Files pane.</p>
<p>Now, you should see the .csv file listed in the File Panel (Panel D) of RStudio.</p>
<p><img src="Images/filepanel.png" id="fig-FilePanel" class="img-fluid"></p>
</section>
<section id="b.-loading-a-.csv-file-into-r" class="level2">
<h2 class="anchored" data-anchor-id="b.-loading-a-.csv-file-into-r">B. Loading a .csv file into R</h2>
<p>Note that while the <code>name.csv</code> file is in the file panel, it has not been loaded into R yet. It is merely present in the folder on the cloud (or your computer). To load the file into R, in the R Script Panel (Panel A), type or paste:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>data_name <span class="ot">=</span> <span class="fu">read.csv</span>(<span class="at">file =</span> <span class="st">"name.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here is a breakdown of the R syntax:</p>
<p><code>data_name</code> is a name of the object (in the form of a dataframe) that we are assigning to the output of <code>read.csv(file = "name.csv")</code> via a <code>=</code> sign.<br>
<br>
<code>read.csv()</code> is a function. It is typically shown in blue.<br>
<br>
You can pass “arguments” inside the <code>()</code> to specific how the function should operate. They are typically shown in various shades of green. A function can have no argument, or many arguments separated by commas (,).<br>
<br>
In read.csv(), the first argument is <code>file</code> and we pass the value <code>"name.csv"</code> to this argument after the <code>=</code>.<br>
<br>
The value of an argument can be a number, logical argement (Ture/Fales), the name of an object in R, or text. WIth some exceptions (e.g., in ggplot), text values should be under a quotation mark (” “) to distinguish it from the name of an object.<br>
<br>
Since R knows the expected sequence of arguments (this is called positional argument), you could skip the argument’s name and just write: <code>data_name = read.csv("name.csv")</code></p>
<p>There are many other arguments in <code>read.csv</code>. To see them, type <code>?read.csv</code> in the Console Panel (Panel B) to display the Help page in the File Console (Panel D). | For now, the default works (e.g., header = TRUE, sep = “,”) because <code>name.csv</code> has a header row and values are separated by “,”. So <code>data_name = read.csv("name.csv")</code> is doing the same opreation as <code>data_name = read.csv("name.csv", header = TRUE, sep = ",")</code>. | Because of this default arguments, R may read a csv file incorrectly if it does not have a header line!</p>
<p>The code above will read <code>name.csv</code> into R and save it as an “object” called <code>data_name</code>. An object is data that is stored in R. Objects are different from files that are being imported into Posit Cloud and RStudio. Objects can be used in the Script or Console Panel (e.g., by running <code>data_name</code>), but files in the File Panel cannot.</p>
<p>Now the object <code>data_name</code> will appear in the Environment Panel (Panel C).</p>
<p><img src="Images/env_panel.png" id="fig-env_panel" class="img-fluid"></p>
<p>In the environment panel, you can click on the object to open it in a new tab in the Script Panel. Note that doing it actually sends a code to the Console Panel as <code>View(data_name)</code></p>
<p>You can show the content of the object in the Console Panel by running (or “calling”) it directly:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>data_name</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  First_name Height_ft Age Academic_standing
1     Denzel       5.8  24         Sophomore
2     Claire       5.0  20            Junior
3       Jack       7.2  29        First-year
4     Quisha       5.7  26         Sophomore
5     Carlos       4.9  24        First-year
6      Julio       5.3  28         Sophomore
7     Hiromi       6.0  26            Senior
8      Jiang       5.5  28            Junior
9       Isha       4.4  18         Sophomore</code></pre>
</div>
</div>
<p>As shown above, when you view an object in the console panel, the file has loaded correctly. The 9 rows and 4 columns are present, as well as the column names.</p>
<p>The object is a <em>dataframe</em>, which stores a two-dimensional table. There are many other classes of R object, see <a href="https://swcarpentry.github.io/r-novice-inflammation/13-supp-data-structures/">here</a> for other classes in R. We will only be using dataframes in this guide.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(data_name)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data.frame"</code></pre>
</div>
</div>
<p>Tips:</p>
<ul>
<li><p>After you type a command (e.g., <code>aov()</code>), you can hit the tab key and RStudio will show the possible arguments.</p></li>
<li><p>RStudio will auto-complete the close bracket.</p></li>
<li><p>If you need to edit your data, the easiest way is to to it in Excel, then upload it to Posit Cloud and read it into R. See the Appendix if you need to edit your data in the .csv file or object within R.</p></li>
</ul>
<hr>
</section>
</section>
<section id="vii.-how-to-generate-descriptive-statistics" class="level1">
<h1>VII. How to generate descriptive statistics</h1>
<p>Let’s say that you measure the length of 12 seashells. Here are the instructions for determining the mean and standard deviation using R:</p>
<p>Follow the instructions in the previous section to enter your continuous data into a new dataframe called <code>seashells</code>. In this case your dependent variable is <code>Length_cm</code> of the seashells.</p>
<p>For example, you can first enter your data into Excel, save it as <code>seashells.csv</code>, and upload that to your project’s directory in Posit.</p>
<p><img src="Images/seashell%20data%20in%20excel.png" id="fig-seashellexcel" class="img-fluid" width="139"></p>
<p>Then you can load it as a dataframe called <code>seashells</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>seashells <span class="ot">=</span> <span class="fu">read.csv</span>(<span class="st">"seashells.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Alternatively, you also run this code to manually enter the data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>seashells <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="st">"Length_cm"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="fl">4.5</span>, <span class="fl">3.1</span>, <span class="fl">4.2</span>, <span class="fl">4.4</span>, <span class="dv">4</span>, <span class="fl">3.6</span>, <span class="fl">3.7</span>, <span class="fl">3.1</span>, <span class="fl">2.8</span>, <span class="fl">4.3</span>, <span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Either way, your data should look like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>seashells</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Length_cm
1        3.0
2        4.5
3        3.1
4        4.2
5        4.4
6        4.0
7        3.6
8        3.7
9        3.1
10       2.8
11       4.3
12       5.0</code></pre>
</div>
</div>
<p>You can check whether Length_cm is continuous (i.e, numeric in R).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(seashells<span class="sc">$</span>Length_cm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "numeric"</code></pre>
</div>
</div>
<p><code>$</code> is an “operator” to get a column from a dataframe’s table.</p>
<p><code>seashells$Length_cm</code> means that from the dataframe <code>seashells</code>, show the column called <code>Length_cm</code>. | | Then, the function <code>class()</code> is applied to the column to examine the “class” of the column, which will typically be “numeric” for a continuous variable and “character” for a categorical variable.</p>
<p>Generate a quick histogram of the data, if you have a continuous variable.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(seashells<span class="sc">$</span>Length_cm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Generate basic summary statistics:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(seashells<span class="sc">$</span>Length_cm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  2.800   3.100   3.850   3.808   4.325   5.000 </code></pre>
</div>
</div>
<p>You can generate the mean, standard deviation (SD) and standard error (SE) using the function <code>summarySE()</code> in the package <code>Rmisc</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summarySE</span>(<span class="at">data =</span> seashells, <span class="at">measurevar =</span> <span class="st">"Length_cm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   .id  N Length_cm        sd        se        ci
1 &lt;NA&gt; 12  3.808333 0.7012435 0.2024316 0.4455489</code></pre>
</div>
</div>
<p>You can also calculate the means and standard deviations of different groups at the same time. Let’s say you measured 12 shells again, but in this experiment you used 3 different species. You can calculate the means and standard deviations of the different species by inputting your data in the following way:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We will add another column to the dataframe</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>seashells<span class="sc">$</span>Species <span class="ot">=</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"Species1"</span>, <span class="dv">4</span>), <span class="fu">rep</span>(<span class="st">"Species2"</span>, <span class="dv">4</span>), <span class="fu">rep</span>(<span class="st">"Species3"</span>, <span class="dv">4</span>))</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co"># To view the data:</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>seashells</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Length_cm  Species
1        3.0 Species1
2        4.5 Species1
3        3.1 Species1
4        4.2 Species1
5        4.4 Species2
6        4.0 Species2
7        3.6 Species2
8        3.7 Species2
9        3.1 Species3
10       2.8 Species3
11       4.3 Species3
12       5.0 Species3</code></pre>
</div>
</div>
<p>Then use <code>summarySE()</code> with an additional groupvars argument:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summarySE</span>(<span class="at">data =</span> seashells, <span class="at">measurevar =</span> <span class="st">"Length_cm"</span>, <span class="at">groupvars =</span> <span class="st">"Species"</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Species N Length_cm        sd        se        ci
1 Species1 4     3.700 0.7615773 0.3807887 1.2118394
2 Species2 4     3.925 0.3593976 0.1796988 0.5718819
3 Species3 4     3.800 1.0295630 0.5147815 1.6382645</code></pre>
</div>
</div>
<p>With this data, you can make a table in Excel to summarize descriptive statistics of seashell lengths from three different species.</p>
<p><img src="Images/Mean%20shell%20length%20by%20species.png" id="fig-meanshell" class="img-fluid"></p>
<hr>
</section>
<section id="viii.-inferential-statistics-graphs" class="level1">
<h1>VIII. Inferential statistics &amp; graphs</h1>
<p><em>Step One: Figure out what statistical test you should use!</em></p>
<p>How do you do that? First, you should be able to clearly write out your experimental question, your hypothesis, and identify your dependent and independent variables.</p>
<p>Once you have your variables identified, determine if they are continuous (measurement) data or nominal (categorical) data. Now you can use the chart below to figure out which type of statistic is right for you. Note that each combination of variable types also has a type of graph or table that is typically used to display the data! Instructions for how to run each stat and make each figure are given in the sections that follow.</p>
<p><img src="Images/Statistic_Chart.png" id="fig-stat" class="img-fluid"></p>
<hr>
<section id="a.-t-test" class="level2">
<h2 class="anchored" data-anchor-id="a.-t-test">A. T-test</h2>
<section id="two-sample-t-test" class="level3">
<h3 class="anchored" data-anchor-id="two-sample-t-test">Two-sample t-test</h3>
<p>A <strong>t-test is used when the independent variable is nominal (categorical) and has exactly two categories (such as an experimental group and a control group) and the dependent variable consists of continuous (measurement) data</strong>. In this case, you are determining if there are differences in the means of the dependent variable between the two categories of the independent variable.</p>
<p>Let’s say that you want to test if a fertilizer has a significant effect on the growth of plants. You measure the height of 10 plants, 5 grown with just soil and 5 grown with soil and added fertilizer. Both groups are grown for the same length of time. Here are the instructions for performing a t-test (for comparing two groups) using R:</p>
<p>According to <a href="#vi.-importing-data-into-r">VI. Importing data into R</a>, import <code>fertilizer.csv</code> from the shared drive (<a href="https://drive.google.com/drive/folders/18p2BJvPVNJhLtYQ-HolEADr5Hwpi67JB?usp=sharing">here</a>) as a daraframe named <code>fertilizer</code>. Also, be sure you are <a href="#loading-packages">Loading packages</a>.</p>
<p>(See <a href="#appendix-3.-create-dataframes-in-r">Appendix 3. Create dataframes in R</a> if you want to create a dataframe manually in R.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This is what the dataframe should look like:</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>fertilizer</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Treatment Plant_height_cm
1             Soil            10.5
2             Soil            13.0
3             Soil             9.0
4             Soil            11.0
5             Soil            14.0
6  Soil+fertilizer            17.0
7  Soil+fertilizer            18.5
8  Soil+fertilizer            20.0
9  Soil+fertilizer            16.0
10 Soil+fertilizer            17.5</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The data has two treatment levels:</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(fertilizer<span class="sc">$</span>Treatment)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Soil"            "Soil+fertilizer"</code></pre>
</div>
</div>
<p>To test whether <code>Plant_height_cm</code> is different between <code>Treatment</code> , in other word, whether <code>Treatment</code> affects <code>Plant_height_cm</code>, <strong>we will use <code>t.test()</code></strong>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(Plant_height_cm <span class="sc">~</span> Treatment, <span class="at">data =</span> fertilizer, <span class="at">alternative =</span> <span class="st">"two.sided"</span>, <span class="at">var.equal =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Two Sample t-test

data:  Plant_height_cm by Treatment
t = -5.6014, df = 8, p-value = 0.0005094
alternative hypothesis: true difference in means between group Soil and group Soil+fertilizer is not equal to 0
95 percent confidence interval:
 -8.893614 -3.706386
sample estimates:
           mean in group Soil mean in group Soil+fertilizer 
                         11.5                          17.8 </code></pre>
</div>
</div>
<p>The argument <code>alternative</code> can also be “greater” or “less” if you have an a priori reason to test for a directional difference between groups.</p>
<p>In this case, the p-value is 0.0005. We reject the null hypothesis that <code>Treatment</code> has no effect on <code>Plant_height_cm</code>. Hence, we can write a summary statement like:</p>
<blockquote class="blockquote">
<p>Plants grown in soil with fertilizer were taller than plants grown soil with no fertilizer (t = -5.601, d.f. = 8, p = 0.0005 or p &lt; 0.001)</p>
</blockquote>
</section>
<section id="visualization" class="level3">
<h3 class="anchored" data-anchor-id="visualization"><strong>Visualization</strong></h3>
<p>Biologists often display the results of an experiment in the form of a graph. There are a few types of graphs encountered in biology: bar graph, boxplot, scatter plot, and line graph. Bar graph and boxplot are useful when your independent variable is categorical and your dependent variable is continuous. Scatter plot and line graph are useful if both variables are continuous. Otherwise, they share many of the same features including a title and labels for the X and Y-axes (including the units of measurement, if any). <strong>The predictor (independent) variable is plotted on the X-axis, while the response (dependent) variable is plotted on the Y-axis.</strong> Refer to your instructor for guideline about writing figure captions.</p>
<p>You can plot the continuous variable from two groups with a bar graph or boxplot.</p>
<section id="bar-graph" class="level4">
<h4 class="anchored" data-anchor-id="bar-graph">Bar graph</h4>
<p>In this example, we will use the dataframe <code>fertilizer</code>, which was used above for the t-test. We will plot the mean <code>Plant_height_cm</code> across different <code>Treatment</code> and add an error bar that represents the standard deviation (or standard error if needed).</p>
<p>Because the dataframe <code>fertilizer</code> does not have the mean or standard deviation, we will use <code>summarySE()</code> to generate the mean, SD, and SE by group, similar to how we generated descriptive statistics in <a href="#vii.-how-to-generate-descriptive-statistics">VII. How to generate descriptive statistics</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We will store the result of summarySE as a new object called height_by_Trt_summary</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>height_by_Trt_summary <span class="ot">=</span> <span class="fu">summarySE</span>(<span class="at">data =</span> fertilizer, </span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">measurevar =</span> <span class="st">"Plant_height_cm"</span>, </span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">groupvars =</span> <span class="st">"Treatment"</span>)</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at height_by_Trt_summary</span></span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>height_by_Trt_summary</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Treatment N Plant_height_cm       sd        se       ci
1            Soil 5            11.5 2.000000 0.8944272 2.483328
2 Soil+fertilizer 5            17.8 1.524795 0.6819091 1.893283</code></pre>
</div>
</div>
<p>We can then use <code>height_by_Trt_summary</code> to build a bar graph.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> height_by_Trt_summary, <span class="fu">aes</span>(<span class="at">x =</span> Treatment, <span class="at">y =</span> Plant_height_cm)) <span class="sc">+</span> </span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">fill =</span> <span class="st">"lightblue"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span> </span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> Plant_height_cm <span class="sc">-</span> sd, <span class="at">ymax =</span> Plant_height_cm <span class="sc">+</span> sd), <span class="at">width =</span> <span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Treatment"</span>, <span class="at">y =</span> <span class="st">"Plant height (cm)"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The <code>ggplot</code> function is from the <code>ggplot2</code> package. It results multiple functions connected by <code>+</code>. The new line are not required, but are used for a better organization of the codes.<br>
<br>
<code>ggplot()</code> specifies a dataframe used for plotting with <code>data = height_by_Trt_summary</code> and the x and y variables to use with <code>aes(x = Treatment, y = Plant_height_cm)</code>.<br>
<br>
<code>geom_bar()</code> adds the bar graph.<br>
The arguments <code>stat="identity"</code> specifies the bar graph to display the values (means, under the column <code>Plant_height_cm</code> in <code>height_by_Trt_summary</code>) as is.<br>
<code>fill="lightblue"</code> and <code>color = "black"</code> controls the color of the box and the color of the outline, respectively.<br>
<br>
<code>geom_errorbar()</code> adds the error bar.<br>
The argument <code>aes(ymin = Plant_height_cm - sd, ymax = Plant_height_cm + sd)</code> specifies the coordinates of the top and bottom of the error bar. In this case, the error bar will begin at (ymin) the mean minus SD and end at (ymax) the mean plus SD).<br>
The argument <code>width = 0.2</code> controls the width of the error bar.<br>
<br>
The last line <code>labs(x = "Treatment", y = "Plant height (cm)")</code> specifies the x and y axis labels.<br>
<br>
You will need at least the first two lines (connected by +) to build a ggplot. For example: <code>ggplot(data = height_by_Trt_summary, aes(x = Treatment, y = Plant_height_cm)) +    geom_bar(stat="identity")</code></p>
<p>See the Appendix if you want to overlay data points on the graph.</p>
</section>
<section id="exporting-graphs" class="level4">
<h4 class="anchored" data-anchor-id="exporting-graphs">Exporting graphs</h4>
<p>To export your plot from Posit Cloud, go to the <code>Plot</code> tab on the Files Panel and click <code>Export</code>.</p>
<p><img src="Images/export_graph.png" class="img-fluid"></p>
<ul>
<li><p>You can simple do <code>Copy to Clipboard</code> and paste the figure into other programs on your computer. This option may not work for all programs and operation systems.</p></li>
<li><p>With <code>Save as Image…</code>, you can adjust the image size so that the font size and image size would fit the document where you intend to use the image for (e.g., lab report, presentation, or poster). You should enter an informative name in the box next to <code>File name:</code>.</p></li>
</ul>
<p>After exporting the figure, it will appear as an image file in the Files Panel in Posit Cloud or as a file in the project’s folder in RStudio on your computer. For Posit Cloud, you should download the image file to your computer.</p>
<p>In the Files Panel, check the box next to the image file that you want to download. Click <code>More</code> and choose <code>Export</code>. In the pop-up window, click <code>Download</code>.</p>
<p><img src="Images/download_graph.png" class="img-fluid" width="542"></p>
</section>
<section id="boxplot" class="level4">
<h4 class="anchored" data-anchor-id="boxplot">Boxplot</h4>
<p>Another good way to display the same data is to use a boxplot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> fertilizer, <span class="fu">aes</span>(<span class="at">x =</span> Treatment, <span class="at">y =</span> Plant_height_cm)) <span class="sc">+</span> </span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Treatment"</span>, <span class="at">y =</span> <span class="st">"Plant height (cm)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Type <code>?geom_boxplot</code> in the Console Panel to show the help page for <code>geom_boxplot()</code>. The values displayed by the hinges and whiskers are explained in the section “Summary statistics”. Note that SD and SE are not displayed in the boxplot.</p>
<p>See the Appendix if you want to overlay data points on the graph.</p>
<hr>
</section>
</section>
</section>
<section id="b.-anova" class="level2">
<h2 class="anchored" data-anchor-id="b.-anova">B. ANOVA</h2>
<section id="global-anova-test" class="level3">
<h3 class="anchored" data-anchor-id="global-anova-test">Global ANOVA test</h3>
<p><strong>An ANOVA is used in a similar circumstance as a t-test: nominal independent variable and continuous dependent variable, but when there are three or more categories for the independent variable (perhaps you tried multiple experimental treatments)</strong>. If this test is significant, you must then perform additional “post hoc” tests to determine exactly which pair(s) of groups are significantly different from one another.</p>
<p>Let’s say you want to measure the effectiveness of penicillin in disrupting <em>E. coli</em> growth. To see if doses of penicillin differ in effectiveness, we will test 3 different concentrations as well as a control (no penicillin). The response variable in this case is measured in CFU (colony forming unit).</p>
<p><strong>Here are the instructions for performing an ANOVA (for comparing more than 2 groups) in R:</strong></p>
<p>According to <a href="#vi.-importing-data-into-r">VI. Importing data into R</a>, import <code>penicillin.csv</code> from the shared drive (<a href="https://drive.google.com/drive/folders/18p2BJvPVNJhLtYQ-HolEADr5Hwpi67JB?usp=sharing">here</a>) as a daraframe named <code>penicillin</code>. Also, be sure you are <a href="#loading-packages">Loading packages</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This is what the dataframe should look like:</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>penicillin</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Penicillin_conc CFU
1             Zero 795
2             Zero 711
3             Zero 703
4             Zero 798
5             Zero 775
6              Low 602
7              Low 696
8              Low 602
9              Low 689
10             Low 709
11          Medium 489
12          Medium 475
13          Medium 411
14          Medium 677
15          Medium 654
16            High   2
17            High   0
18            High   2
19            High   1
20            High   0</code></pre>
</div>
</div>
<p>We need to adjust the column Penicillin_conc to be a ordered factor. If not, ggplot will arrange it alphabetically.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>penicillin<span class="sc">$</span>Penicillin_conc <span class="ot">=</span> <span class="fu">factor</span>(penicillin<span class="sc">$</span>Penicillin_conc, <span class="at">ordered =</span> T, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Zero"</span>, <span class="st">"Low"</span>, <span class="st">"Medium"</span>, <span class="st">"High"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To test whether <code>Penicillin_conc</code> affects <code>CFU</code>, we will build the model using <code>aov()</code>, then use <code>summary()</code> to show the full results:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>p_anova_model <span class="ot">=</span> <span class="fu">aov</span>(<span class="at">formula =</span> CFU <span class="sc">~</span> Penicillin_conc, <span class="at">data =</span> penicillin)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(p_anova_model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                Df  Sum Sq Mean Sq F value   Pr(&gt;F)    
Penicillin_conc  3 1707374  569125   121.5 3.21e-11 ***
Residuals       16   74973    4686                     
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>Note that the results are similar to the F-test for linear regressions, because ANOVA and regression are both linear models. But in ANOVA, we don’t have predictions of slope and intercepts.</p>
<p>The p value is shown below <code>Pr(&gt;F)</code>. In this case, since the p-value (3.21e-11) is &lt;0.05, we reject the null hypothesis that <code>Penicillin_conc</code> has no effect on <code>CFU</code> Therefore, we can write a summary statement to say:</p>
<blockquote class="blockquote">
<p>Penicillin concentration affects CFU in <em>E. coli</em> (F = 121.5, d.f. = 3 and 16, p &lt; 0.0001)</p>
</blockquote>
</section>
<section id="post-hoc-test" class="level3">
<h3 class="anchored" data-anchor-id="post-hoc-test"><strong><em>Post hoc test</em></strong></h3>
<p><u><em>If the p value is below 0.05, you must run post hoc tests to compare the means.</em></u> We can compare all means using <strong>Tukey HSD</strong> to obtain information about which pair(s) of penicillin concentrations are significantly different from one another.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>hsd <span class="ot">=</span> <span class="fu">HSD.test</span>(p_anova_model, <span class="at">trt =</span> <span class="st">"Penicillin_conc"</span>)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>hsd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$statistics
   MSerror Df   Mean       CV      MSD
  4685.825 16 489.55 13.98286 123.8637

$parameters
   test          name.t ntr StudentizedRange alpha
  Tukey Penicillin_conc   4         4.046093  0.05

$means
         CFU       std r       se Min Max Q25 Q50 Q75
High     1.0   1.00000 5 30.61315   0   2   0   1   2
Low    659.6  53.06882 5 30.61315 602 709 602 689 696
Medium 541.2 117.49979 5 30.61315 411 677 475 489 654
Zero   756.4  46.04129 5 30.61315 703 798 711 775 795

$comparison
NULL

$groups
         CFU groups
Zero   756.4      a
Low    659.6     ab
Medium 541.2      b
High     1.0      c

attr(,"class")
[1] "group"</code></pre>
</div>
</div>
<p>Under <code>$groups</code>, the letters indicate which groups are significantly different from each other in a post hoc test. Groups with the same letter are not significantly different from each other (e.g., Zero and Low, which both have “a”), while groups with different letters are significantly different (e.g., Medium and High, which has “b” and “c”, respectively).</p>
</section>
<section id="visualization-1" class="level3">
<h3 class="anchored" data-anchor-id="visualization-1"><strong>Visualization</strong></h3>
<section id="bar-graph-1" class="level4">
<h4 class="anchored" data-anchor-id="bar-graph-1"><strong>Bar graph</strong></h4>
<p>Similar to a t-test, we can create a bar graph with error bars after using <code>summarySE()</code> to generate the mean, SD, and SE for each predictor group.</p>
<p>After you export the figure, you can use Words, PowerPoints, Preview (Macs), or other software to add the post hoc letter labels from <code>hsd</code>. Alternatively, you can do this using codes in R. See Appendix 2 Advanced graphing.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>penicillin_summary <span class="ot">=</span> <span class="fu">summarySE</span>(<span class="at">data =</span> penicillin, <span class="at">measurevar =</span> <span class="st">"CFU"</span>, <span class="at">groupvars =</span> <span class="st">"Penicillin_conc"</span>)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> penicillin_summary, <span class="fu">aes</span>(<span class="at">x =</span> Penicillin_conc, <span class="at">y =</span> CFU)) <span class="sc">+</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>) <span class="sc">+</span> </span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin=</span>CFU<span class="sc">-</span>sd, <span class="at">ymax=</span>CFU<span class="sc">+</span>sd), <span class="at">width=</span>.<span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Penicillin concentration"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Your final figure should include the post hoc letters, similar to the figure below. You can export the graph to add the letters manually. Alternatively, you can do it in R. See <a href="#adding-post-hoc-grouping-into-bar-graph-and-boxplot">Adding post hoc grouping into bar graph and boxplot</a>.</p>
<div class="cell">
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="box-plot" class="level4">
<h4 class="anchored" data-anchor-id="box-plot"><strong>Box plot</strong></h4>
<p>Similar to the t-test, we can plot the data used for ANOVA using a boxplot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> penicillin, <span class="fu">aes</span>(<span class="at">x =</span> Penicillin_conc, <span class="at">y =</span> CFU)) <span class="sc">+</span> </span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Penicillin concentration"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<hr>
</section>
</section>
</section>
<section id="c.-linear-regression" class="level2">
<h2 class="anchored" data-anchor-id="c.-linear-regression">C. Linear regression</h2>
<section id="simple-linear-regression" class="level3">
<h3 class="anchored" data-anchor-id="simple-linear-regression">Simple linear regression</h3>
<p>A linear regression is employed to determine if there is a relationship between two variables, X and Y <strong>when both variables are continuous</strong> (numerical or measurement) type data. In other words, whether the slope of the linear regression line is different from zero (remember a flat line has a slope of zero). This statistical test will also indicate the direction and magnitude of the relationship. The direction is inferred from the slope of the line (m), while the magnitude or strength of the relationship is inferred from the coefficient of determination (R2). If R2 = 1, then the line is a perfect “fit,” meaning that Y can be predicted from X (or vice versa) using the equation of the line (y = mx + b). If R2 &lt; 1, then the line is not a perfect “fit,” meaning that only a proportion of Y can be predicted from X (or vice versa) using the equation of the line (y = mx + b). For example, if R2 = 0.94, then 94% of the variation in Y can be explained by X. Note that the equation of the line is ONLY relevant if there is a significant relationship (p &lt; 0.05).</p>
<p>Let’s say that you want to determine if the chirp rates of the snowy tree cricket are related to temperature. You measure the temperature and the rate of 10 chirping crickets at those respective temperatures. Since crickets are ectotherms and their metabolism is affected by their external environment, we should expect to see an increase in chirping as temperature increases.</p>
<p>According to <a href="#vi.-importing-data-into-r">VI. Importing data into R</a>, import <code>chirps.csv</code> from the shared drive (<a href="https://drive.google.com/drive/folders/18p2BJvPVNJhLtYQ-HolEADr5Hwpi67JB?usp=sharing">here</a>) as a daraframe named <code>chirps</code>. Also, be sure you are <a href="#loading-packages">Loading packages</a>.</p>
<p>We have two continuous variables <code>temperature</code> and <code>chips_per_min</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Display the dataframe chirps</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>chirps</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   temperature chirps_per_min
1           31             79
2           22             64
3           34             79
4           29             73
5           27             69
6           24             62
7           21             59
8           29             70
9           26             64
10          29             68</code></pre>
</div>
</div>
<p>In R the <code>lm</code> function can fit a number of linear models, including linear regression.</p>
<p>The code below creates a linear model (lm_CT) to predict <code>chirps_per_min</code> (the response or dependent variable) by <code>temperature</code> (the predictor or independent variable):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>lm_CT <span class="ot">=</span> <span class="fu">lm</span>(<span class="at">formula =</span> chirps_per_min <span class="sc">~</span> temperature, <span class="at">data =</span> chirps)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can view results of the linear regression using <code>summary()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(lm_CT)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = chirps_per_min ~ temperature, data = chirps)

Residuals:
    Min      1Q  Median      3Q     Max 
-3.5000 -1.6667 -0.1667  1.2778  4.3889 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)  26.3889     6.1375   4.300 0.002617 ** 
temperature   1.5556     0.2234   6.962 0.000117 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 2.714 on 8 degrees of freedom
Multiple R-squared:  0.8583,    Adjusted R-squared:  0.8406 
F-statistic: 48.47 on 1 and 8 DF,  p-value: 0.000117</code></pre>
</div>
</div>
<p>Here is an explanation of the output:</p>
<p>Under “Coefficients”, it shows a table with two rows corresponding to (1) the intercept and (2) the slope (of the predictor variable, i.e., temperature) and four columns:</p>
<ul>
<li><p><strong>Estimate: the estimated intercept and slope of the linear regression line.</strong></p></li>
<li><p>Std. Error: Standard error of the estimate</p></li>
<li><p>t value: test statistics from a t test of whether the estimate is significantly different from zero.</p></li>
<li><p><strong>Pr(&gt;|t|): P value of the t test</strong></p></li>
</ul>
<p>In the last three lines, it shows:</p>
<ul>
<li><p>Residual standard error: lower values suggest better fit (i.e., less distance between the data point and the predicted line).</p></li>
<li><p><strong>Degrees of freedom: the maximum number of logically independent values.</strong></p></li>
<li><p><strong>Multiple R-squared: the percentage of the variation in the response (dependent) variable that the model (dependent variables) explains.</strong></p></li>
<li><p>Adjusted R-squared*: R-squared adjusted for the number of terms in the model. Use this to compare models with more than one predictor variable.</p></li>
<li><p>F-statistics: Statistics of a F-test that evaluates the fit of the model.</p></li>
</ul>
<p>*If there are multiple predictor variables, although the F-test is at the end of the statistical output, it should be examined first to determine whether the linear model fits the data. Only when the model is well fitted (as judged by the F- and P-values), should you further interpret the significance of the slope of a particular predictor.</p>
<p>In this example, the linear regression is significant because the p-value of the F test is lower than 0.05 (p = 0.000117), so we can continue to examine the intercept and slope.</p>
<p>The slope is 1.5556 (under “Estimate” in “Coefficient”) and the t-test indicates that it is significantly different from zero (p = 0.000117). Note that this p-value is the same as that of the F-test only when there is a single predictor variable. In multiple regression, there will be separate t-tests for each predictor variable.</p>
<p>Therefore, we can write a statement to summarize the result and provide statistical support:</p>
<blockquote class="blockquote">
<p><strong>There is a positive relationship between temperature and chirp rate (slope =</strong> 1.556<strong>, F = 48.47, d.f. = 1 and 8, p =</strong> 0.0001<strong>).</strong></p>
</blockquote>
</section>
<section id="visualization-2" class="level3">
<h3 class="anchored" data-anchor-id="visualization-2">Visualization</h3>
<section id="scatter-plot-line-plot" class="level4">
<h4 class="anchored" data-anchor-id="scatter-plot-line-plot">Scatter plot / Line plot</h4>
<p>For the dataframe <code>chirps</code>, we will plot <code>temperature</code> as the predictor (independent) variable on the X axis and chirps_per_min as the response (dependent) variable on the Y axis. A linear regression line will be added.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> chirps, <span class="fu">aes</span>(<span class="at">x =</span> temperature, <span class="at">y =</span> chirps_per_min)) <span class="sc">+</span> </span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> lm, <span class="at">se =</span> F, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Chirps per minute"</span>, <span class="at">x =</span> <span class="fu">expression</span>(<span class="st">"Temperature "</span> ( degree<span class="sc">*</span>C)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-48-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><u>Always do the following (regardless of the result):</u> Make sure your axes are labeled properly in Title Case and include units of measure.</p>
<p><u>If the p-value is significant, do the following:</u> show the linear fit on the graph (as shown above).</p>
<p><u>If the p-value is not significant:</u> this means the line fit is not meaningful, so just show the scatter plot without the linear fit (as shown below).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> chirps, <span class="fu">aes</span>(<span class="at">x =</span> temperature, <span class="at">y =</span> chirps_per_min)) <span class="sc">+</span> </span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Chirps per minute"</span>, <span class="at">x =</span> <span class="fu">expression</span>(<span class="st">"Temperature "</span> ( degree<span class="sc">*</span>C)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-49-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Some instructors or journals prefer to display an insignificant trend with a dashed line:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> chirps, <span class="fu">aes</span>(<span class="at">x =</span> temperature, <span class="at">y =</span> chirps_per_min)) <span class="sc">+</span> </span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> lm, <span class="at">se =</span> F, <span class="at">color =</span> <span class="st">"grey"</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Chirps per minute"</span>, <span class="at">x =</span> <span class="fu">expression</span>(<span class="st">"Temperature "</span> ( degree<span class="sc">*</span>C)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-50-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<hr>
</section>
</section>
</section>
<section id="d.-two-types-of-chi-square-x2-tests" class="level2">
<h2 class="anchored" data-anchor-id="d.-two-types-of-chi-square-x2-tests">D. Two types of Chi-square (X<sup>2</sup>) tests</h2>
<p><strong>Chi-Square tests are used when both the dependent and independent variables are nominal or categorical.</strong></p>
<section id="chi-square---goodness-of-fit-test" class="level3">
<h3 class="anchored" data-anchor-id="chi-square---goodness-of-fit-test"><strong>Chi-square - Goodness of Fit test</strong></h3>
<p>The <strong>Goodness of Fit test</strong> is used to determine if the observed result (i.e.&nbsp;collected during an experiment) is significantly different than the “expected” result (i.e.&nbsp;that based on the laws of probability). It involves calculating the degrees of freedom (DF), which corresponds to the following equation: DF = (C – 1) where C is the number of categories. It also involves calculating the Chi-square value (X<sup>2</sup>), which corresponds to the following equation: X<sup>2</sup> = <span class="math inline">\(\sum{D^2/E}\)</span> where D is the difference between the observed result (O) and the expected result (E). The biologist determines if the X<sup>2</sup> value is less than the “critical” X<sup>2</sup> value for an experiment with a particular DF to determine if there is a significant difference (p value &lt; 0.05) between the observed and expected result.</p>
<p><img src="Images/chisq.png" id="chisq" class="img-fluid" alt="Chi-square">.</p>
<p>Let’s say that you have crustaceans and test their preference for a light or dark environment. In this example there is only ONE variable: the observed preference (are they in the light or dark when you check on them?). You want to see if the distribution of crustaceans is random (roughly 50% in the light and 50% in the dark because they have no preference and distribute according to the laws of probability) OR if the distribution is non-random because they prefer either dark or light.</p>
<p>The <strong>null hypothesis</strong> for this test is that the variable of interest (light/dark environment) will have no impact on the distribution of the data. In other words, the null hypothesis assumes a random distribution, which means the <strong>observed is no different from the expected</strong>. If p&gt;0.05, you accept the null.</p>
<p>A <strong>significant result</strong> (p&lt;0.05) means the distribution of your <strong>observed data is different than expected</strong>; it is non-random.</p>
<p>The goodness of fit Chi-square test is used to test whether the distribution of a categorical variable is significantly different than an “expected” distribution.</p>
<p>For example, in a species of crustacean, we counted the number of individuals that prefer a light and a dark treatment. We may expect the light:dark ratio to be 1:1 and we can test it with the chi-square goodness of fit test.</p>
<p>According to <a href="#vi.-importing-data-into-r">VI. Importing data into R</a>, import <code>crust_sp1.csv</code> from the shared drive (<a href="https://drive.google.com/drive/folders/18p2BJvPVNJhLtYQ-HolEADr5Hwpi67JB?usp=sharing">here</a>) as a daraframe named <code>crust_sp1</code>. Also, be sure you are <a href="#loading-packages">Loading packages</a>.</p>
<p>We will create a :</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This file may be similar to how your raw data will look in a “long form”</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>crust_sp1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Crustacean Result
1   Species 1  light
2   Species 1  light
3   Species 1   dark
4   Species 1  light
5   Species 1  light
6   Species 1  light
7   Species 1  light
8   Species 1  light
9   Species 1   dark
10  Species 1   dark
11  Species 1  light
12  Species 1  light
13  Species 1  light
14  Species 1  light
15  Species 1  light
16  Species 1  light
17  Species 1  light
18  Species 1   dark
19  Species 1   dark
20  Species 1   dark
21  Species 1  light
22  Species 1  light
23  Species 1  light
24  Species 1  light</code></pre>
</div>
</div>
<p>We can convert this into the “short form” :</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>crust_table <span class="ot">=</span> <span class="fu">table</span>(crust_sp1[,<span class="st">"Result"</span>])</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>crust_table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 dark light 
    6    18 </code></pre>
</div>
</div>
<p>We can then test whether the observed distribution of <code>Result</code> is equally distributed among the two groups (i.e., 1:1) using <code>chisq.test()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">chisq.test</span>(<span class="at">x =</span> crust_table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Chi-squared test for given probabilities

data:  crust_table
X-squared = 6, df = 1, p-value = 0.01431</code></pre>
</div>
</div>
<p>The X-squared value (6) is the “Pearson” test chi-square value. It is used, with the degrees of freedom (df) value, to obtain the p-value from a chi-squared table.</p>
<p>The null hypothesis of the test is that the groups are equally distributed. The p-value is &lt; 0.05, which means that we can reject the null hypothesis. Therefore, our observed distribution of light preferences by crustaceans is significantly different from 1:1.</p>
<section id="visualization-3" class="level4">
<h4 class="anchored" data-anchor-id="visualization-3"><strong>Visualization</strong></h4>
<p>There are several ways of visually representing X<sup>2</sup> results in either a figure or a table. Ask your instructor about their preference.</p>
<p><em>For a table</em>:</p>
<p>The table should be <u>a simplified version</u> of the table R used to run the analysis.</p>
<p><img src="Images/chisq_table.png" id="fig-chisq_table" class="img-fluid"></p>
<p><em>For a graph</em>:</p>
<p>The type of graph used to display frequency (count) data is called a histogram.</p>
<p>You can plot the data for a goodness-of-fit Chi-square test by using the raw data in <code>crust_sp1</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> crust_sp1, <span class="fu">aes</span>(<span class="at">x =</span> Result)) <span class="sc">+</span> </span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Result"</span>, <span class="at">y =</span> <span class="st">"Count"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-55-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Alternative, you could also plot using the count data <code>crust_table</code> after converting it to a dataframe:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">data.frame</span>(crust_table), <span class="fu">aes</span>(Var1, Freq, <span class="at">fill =</span> Var1)) <span class="sc">+</span> </span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>() <span class="sc">+</span> </span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Result"</span>, <span class="at">y =</span> <span class="st">"Count"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-56-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fill tells R to color the bar by Var1</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="co"># you can remove the side legend by adding:</span></span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a><span class="co"># + theme(legend.position = "none")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="chi-square---contingency-test" class="level3">
<h3 class="anchored" data-anchor-id="chi-square---contingency-test">Chi-square - Contingency test</h3>
<p>The <strong>Contingency test</strong>, also called the <strong>Test of Independence</strong>, is used to determine if the frequency of outcomes for one set of observations matches those for another set of observations.</p>
<p>For example, let’s imagine that we want to know whether two different species of crustaceans exhibit the same preference for light or dark. In other words, is the preference for light or dark contingent on which species it is? To find this out one uses a version of a Chi-square test known as a contingency table.</p>
<p>In this case, your <strong>independent variable</strong> is species and your <strong>dependent variable</strong> is preference (because you think preference might depend on species); your null hypothesis is that there is no difference in the preference for light versus dark between Species 1 and 2. When you run a contingency test you produce a X<sup>2</sup> value as in the Goodness of Fit test. If the X<sup>2</sup> value is large enough (leading to p&lt;0.05), then you can reject the null hypothesis that the proportion of crustaceans showing a preference for light is <strong>not different</strong> between the two species, and state that the data shows that the preference for light or dark <strong>is contingent</strong> on the species. Another way to say the same thing: the two species show significantly different preferences for light.</p>
<p>Let’s imagine that we test the preference of two crustacean species for light versus dark and compare the preferences. The values could be arranged in a table like the one below.</p>
<p><img src="Images/Contingency_data.png" id="fig-Contingency_data" class="img-fluid"></p>
<p>If the preferences of the two species were the same, they should each have roughly the same frequencies as the pooled results, which show animals split between dark and light, with possibly a small preference for dark (28 dark and 20 light). Do species 1 and 2 appear to have similar preferences? Note that the total number of observations/ occurrences does not have to be the same for both species. A contingency test will correct for different sizes between samples.</p>
<p>We will begin by making up the data for species 2:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>crust_sp2 <span class="ot">=</span>  <span class="fu">data.frame</span>(<span class="at">Crustacean =</span> <span class="st">"Species 2"</span>, <span class="at">Result =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"dark"</span>, <span class="dv">22</span>), <span class="fu">rep</span>(<span class="st">"light"</span>, <span class="dv">2</span>)))</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>crust_sp2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Crustacean Result
1   Species 2   dark
2   Species 2   dark
3   Species 2   dark
4   Species 2   dark
5   Species 2   dark
6   Species 2   dark
7   Species 2   dark
8   Species 2   dark
9   Species 2   dark
10  Species 2   dark
11  Species 2   dark
12  Species 2   dark
13  Species 2   dark
14  Species 2   dark
15  Species 2   dark
16  Species 2   dark
17  Species 2   dark
18  Species 2   dark
19  Species 2   dark
20  Species 2   dark
21  Species 2   dark
22  Species 2   dark
23  Species 2  light
24  Species 2  light</code></pre>
</div>
</div>
<p>Then we will add it to the previous <code>crust_sp1</code> to make a new object:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>crust_twoSp <span class="ot">=</span> <span class="fu">rbind</span>(crust_sp1, crust_sp2)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>crust_twoSp</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a><span class="co"># The table is very long; so we are not displaying it here.</span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a><span class="co"># rbind() combines two dataframes by adding the second datafrome (crust_sp2) to the bottom on the first dataframe (crust_sp1).</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will first reshape our data into a contingency table using <code>table()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>crust_twoSp_table <span class="ot">=</span> <span class="fu">table</span>(crust_twoSp<span class="sc">$</span>Crustacean, crust_twoSp<span class="sc">$</span>Result)</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>crust_twoSp_table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           
            dark light
  Species 1    6    18
  Species 2   22     2</code></pre>
</div>
</div>
<p>We can then test the null hypothesis that there is no difference in the preference for light versus dark between Species 1 and 2.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">chisq.test</span>(crust_twoSp_table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Pearson's Chi-squared test with Yates' continuity correction

data:  crust_twoSp_table
X-squared = 19.286, df = 1, p-value = 1.125e-05</code></pre>
</div>
</div>
<section id="visualization-4" class="level4">
<h4 class="anchored" data-anchor-id="visualization-4"><strong>Visualization</strong></h4>
<p>You can plot the contingency table by <code>geom_bar()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> crust_twoSp, <span class="fu">aes</span>(<span class="at">x =</span> Result)) <span class="sc">+</span> <span class="fu">geom_bar</span>() <span class="sc">+</span> <span class="fu">facet_grid</span>(<span class="sc">~</span>Crustacean) <span class="sc">+</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Result"</span>, <span class="at">y =</span> <span class="st">"Count"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-61-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<hr>
</section>
</section>
</section>
</section>
<section id="appendix" class="level1">
<h1>Appendix</h1>
<section id="appendix-1.-editing-data-in-r" class="level2">
<h2 class="anchored" data-anchor-id="appendix-1.-editing-data-in-r">Appendix 1. Editing data in R</h2>
<p>Some times you would need to edit your data after loading it into R. There are two options.</p>
<ol type="1">
<li><p>(Beginner) Modify the .csv file that was uploaded to Posit.</p>
<ul>
<li><p>Click on the .csv file in the <code>Files</code> panel and click <code>View File</code>.</p></li>
<li><p>It will open the .csv file in the top left panel.</p></li>
<li><p>You can modify any values there and save it.</p></li>
<li><p><strong>Once the editing is done, you must load the file into R again using read.csv().</strong></p></li>
<li><p>Note: If you are using Posit Cloud, this option modifies the .csv file on the cloud, but the file on your computer is NOT modified.</p></li>
</ul></li>
<li><p>(Advanced) Modify the dataframe that was loaded into R.<br>
<br>
It is a good practice to keep a record of all editing in your R script and to not modify the raw data file. To do this, we will modify the dataframe that is already loaded into R, using codes that will be saved in the R script. Keep reading if you want to try this option.</p></li>
</ol>
<p>After loading the data into R as a data frame, you can modify it using the command line.</p>
<p>We will use <code>test_df</code> as an example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a dataframe called test_df using the command data.frame()</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>test_df <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">Pred =</span> <span class="fu">c</span>(<span class="st">"A"</span>,<span class="st">"A"</span>, <span class="st">"B"</span>,<span class="st">'B'</span>), <span class="at">Response =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>)</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Notes: </span></span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Inside data.frame(), we provide the name of a column with "Pred = " or "Response= " and the value of each row with c("A","A", "B",'B') or 1:4.</span></span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 1:4 is the same as c(1, 2, 3, 4)</span></span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a><span class="co"># For non-numeric values inside c(), we need to add "" around the characters.</span></span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Display the dataframe test_df</span></span>
<span id="cb67-11"><a href="#cb67-11" aria-hidden="true" tabindex="-1"></a>test_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Pred Response
1    A        1
2    A        2
3    B        3
4    B        4</code></pre>
</div>
</div>
<section id="changing-column-names" class="level3">
<h3 class="anchored" data-anchor-id="changing-column-names">Changing column names</h3>
<p><strong>Option 1.</strong> You can change the column name after knowing the index (position) of the name that needs to be changed.</p>
<p>For example, we want to change <code>Pred</code> to <code>Predictor</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Display the column names of test_df to check the index of Pred</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Index refer to the position of an element in a vector (or list)</span></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(test_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Pred"     "Response"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Since "Pred" is the first element in names(test_df), we can use the index in the next command to change it.</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(test_df)[<span class="dv">1</span>] <span class="ot">=</span> <span class="st">"Predictor"</span></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Display test_df to check</span></span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>test_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Predictor Response
1         A        1
2         A        2
3         B        3
4         B        4</code></pre>
</div>
</div>
<p><strong>Option 2.</strong> We can write one line to change the name without knowing what index it is. It’s useful if you have a larger data frame with lots of columns.</p>
<p>For example, we want to change <code>Response</code> to <code>Values</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(test_df)[<span class="fu">names</span>(test_df)<span class="sc">==</span><span class="st">"Response"</span>] <span class="ot">=</span> <span class="st">"Values"</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>test_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Predictor Values
1         A      1
2         A      2
3         B      3
4         B      4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Breaking down the one-liner:</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a><span class="co"># names(test_df)=="Response" # gives a True/False logical vector</span></span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a><span class="co"># names(test_df)[...] # call what's inside the square bracket</span></span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a><span class="co"># names(test_df)[names(test_df)=="Response"] # So this calls what's TRUE in the logical vector, which is the name "Response".</span></span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a><span class="co"># The last command (= "Values" ) assign a new name for what's on the left side of the equal sign.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are many other ways to do it. You can explore on your own!</p>
</section>
<section id="changing-values-in-a-data-frame" class="level3">
<h3 class="anchored" data-anchor-id="changing-values-in-a-data-frame">Changing values in a data frame</h3>
<p>This is best done by knowing the index of the column and row of the cell that needs to be changed.</p>
<p>For example, if we want to change the 3 to 300 in test_df:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can first take a look at the dataframe</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(test_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Since we know that it's in the third row and the second column, we can do this:</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>test_df[<span class="dv">3</span>,<span class="dv">2</span>] <span class="ot">=</span> <span class="dv">300</span></span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>test_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Predictor Values
1         A      1
2         A      2
3         B    300
4         B      4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Note that indexing for dataframe goes from row to column in R</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a><span class="co"># R start counting from 1. (While other programming language, like Python, starts at 0.)</span></span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a><span class="co"># So [3,2] means row 3 and column 2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="combining-text-from-two-columns" class="level3">
<h3 class="anchored" data-anchor-id="combining-text-from-two-columns">Combining text from two columns</h3>
<p>For example, we have a dataframe called <code>test_df2</code> and want to combine the columns <code>Variable1</code> and <code>Variable2</code> into a new column.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create test_df2</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>test_df2 <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">Variable1 =</span> <span class="fu">c</span>(<span class="st">"Cold"</span>,<span class="st">"Cold"</span>,<span class="st">'Hot'</span>,<span class="st">"Hot"</span>),</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">Variable2 =</span> <span class="fu">c</span>(<span class="st">"North"</span>,<span class="st">"East"</span>,<span class="st">'South'</span>,<span class="st">"West"</span>),</span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">Response =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>)</span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Display test_df2. It has three columns.</span></span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a>test_df2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Variable1 Variable2 Response
1      Cold     North        1
2      Cold      East        2
3       Hot     South        3
4       Hot      West        4</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a new column called Treatment in test_df2</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Do that by pasting the values from each row of Variable1 and Variable1, and separating them with " and ".</span></span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Note that the spaces flanking "and" is needed.</span></span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>test_df2<span class="sc">$</span>Treatment <span class="ot">=</span> <span class="fu">paste</span>(test_df2<span class="sc">$</span>Variable1, test_df2<span class="sc">$</span>Variable2, <span class="at">sep =</span> <span class="st">" and "</span>) </span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Note: avoid using sep = "&amp;" as it may cause an error!</span></span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Display test_df2. It should have the fourth column.</span></span>
<span id="cb82-8"><a href="#cb82-8" aria-hidden="true" tabindex="-1"></a>test_df2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Variable1 Variable2 Response      Treatment
1      Cold     North        1 Cold and North
2      Cold      East        2  Cold and East
3       Hot     South        3  Hot and South
4       Hot      West        4   Hot and West</code></pre>
</div>
</div>
<hr>
</section>
</section>
<section id="appendix-2.-advanced-graphing" class="level2" data-link="How to graph in R">
<h2 data-link="How to graph in R" class="anchored" data-anchor-id="appendix-2.-advanced-graphing">Appendix 2. Advanced graphing</h2>
<section id="overlaying-data-points-in-bar-graph-and-boxplot" class="level3">
<h3 class="anchored" data-anchor-id="overlaying-data-points-in-bar-graph-and-boxplot">Overlaying data points in bar graph and boxplot</h3>
<p>When visualizing data, it is sometime beneficial to show the raw data. We can overlap the data points on a bar graph. This example uses the raw data <code>fertilizer</code> and the dataframe <code>height_by_Trt_summary</code> generated by <code>summarySE</code> from <a href="#a.-t-test">A. T-test</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> height_by_Trt_summary, <span class="fu">aes</span>(<span class="at">x =</span> Treatment, <span class="at">y =</span> Plant_height_cm)) <span class="sc">+</span> </span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">fill =</span> <span class="st">"lightblue"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span> </span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> Plant_height_cm <span class="sc">-</span> sd, <span class="at">ymax =</span> Plant_height_cm <span class="sc">+</span> sd), <span class="at">width =</span> <span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Treatment"</span>, <span class="at">y =</span> <span class="st">"Plant height (cm)"</span>) <span class="sc">+</span></span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">data =</span> fertilizer)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-70-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="co"># It requires adding a new argument geom_point(), where the arguments specify getting the raw data from fertilizer and "jittering" the points to avoid overlapping points with the same values. </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can do the same for a boxplot. When an outlier is present, it will be displayed as a point by <code>geom_boxplot()</code>. Therefore, the argument <code>outlier.colour = "grey"</code> is used to specify the color of the outlier point.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> fertilizer, <span class="fu">aes</span>(<span class="at">x =</span> Treatment, <span class="at">y =</span> Plant_height_cm)) <span class="sc">+</span> </span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">outlier.colour =</span> <span class="st">"grey"</span>) <span class="sc">+</span></span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Treatment"</span>, <span class="at">y =</span> <span class="st">"Plant height (cm)"</span>) <span class="sc">+</span></span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">width =</span> <span class="fl">0.1</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-71-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co"># There is no outlier point in this example!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="adding-post-hoc-grouping-into-bar-graph-and-boxplot" class="level3">
<h3 class="anchored" data-anchor-id="adding-post-hoc-grouping-into-bar-graph-and-boxplot">Adding post hoc grouping into bar graph and boxplot</h3>
<p>We will use the dataframe <code>penicillin</code>, <code>penicillin_summary</code>, and <code>hsd</code> created in <a href="#b.-anova">B. ANOVA</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can use the letters from hsd to build a dataframe to help us label the plot.</span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a><span class="co"># temp_label will have the data from hsd$means and the group letters from hsd$groups. </span></span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>temp_label <span class="ot">=</span> <span class="fu">merge</span>(hsd<span class="sc">$</span>means, hsd<span class="sc">$</span>groups, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"row.names"</span>, <span class="fu">names</span>(hsd<span class="sc">$</span>means)[<span class="dv">1</span>]))</span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a><span class="co"># We need to have a column called Penicillin_conc so the ggplot can refer to this column. Instead of manually specifying the column name, we will retrieve it from hsd$parameters$name.t </span></span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-8"><a href="#cb88-8" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(temp_label)[<span class="dv">1</span>] <span class="ot">=</span> hsd<span class="sc">$</span>parameters<span class="sc">$</span>name.t </span>
<span id="cb88-9"><a href="#cb88-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-10"><a href="#cb88-10" aria-hidden="true" tabindex="-1"></a>temp_label</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Penicillin_conc   CFU       std r       se Min Max Q25 Q50 Q75 groups
1            High   1.0   1.00000 5 30.61315   0   2   0   1   2      c
2             Low 659.6  53.06882 5 30.61315 602 709 602 689 696     ab
3          Medium 541.2 117.49979 5 30.61315 411 677 475 489 654      b
4            Zero 756.4  46.04129 5 30.61315 703 798 711 775 795      a</code></pre>
</div>
</div>
<p>Bar plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Then, geom_text takes data from temp_label to add the post hoc letters to the plot.  </span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a><span class="co"># y = Max + max(Max)*0.1 calculates a y-coordinate to place the label based on the Max values of the data.</span></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Jittered data point is included here for your reference.</span></span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> penicillin_summary, <span class="fu">aes</span>(<span class="at">x =</span> Penicillin_conc, <span class="at">y =</span> CFU)) <span class="sc">+</span></span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>) <span class="sc">+</span> </span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin=</span>CFU<span class="sc">-</span>sd, <span class="at">ymax=</span>CFU<span class="sc">+</span>sd), <span class="at">width=</span>.<span class="dv">2</span>,</span>
<span id="cb90-8"><a href="#cb90-8" aria-hidden="true" tabindex="-1"></a>                 <span class="at">position=</span><span class="fu">position_dodge</span>(.<span class="dv">9</span>)) <span class="sc">+</span></span>
<span id="cb90-9"><a href="#cb90-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Penicillin concentration"</span>) <span class="sc">+</span> </span>
<span id="cb90-10"><a href="#cb90-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">data =</span> penicillin, <span class="fu">aes</span>(<span class="at">x =</span> Penicillin_conc, <span class="at">y =</span> CFU), <span class="at">width =</span> <span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb90-11"><a href="#cb90-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text</span>(<span class="at">data =</span> temp_label, <span class="fu">aes</span>(<span class="at">x =</span> Penicillin_conc, </span>
<span id="cb90-12"><a href="#cb90-12" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">y =</span> Max <span class="sc">+</span> <span class="fu">max</span>(Max)<span class="sc">*</span><span class="fl">0.1</span>, </span>
<span id="cb90-13"><a href="#cb90-13" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">label =</span> groups))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-73-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Boxplot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> penicillin, <span class="fu">aes</span>(<span class="at">x =</span> Penicillin_conc, <span class="at">y =</span> CFU)) <span class="sc">+</span> </span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">width =</span> <span class="fl">0.1</span>, <span class="at">height =</span> <span class="dv">0</span>) <span class="sc">+</span> </span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Penicillin concentration"</span>) <span class="sc">+</span> </span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data =</span> temp_label, <span class="fu">aes</span>(<span class="at">x =</span> Penicillin_conc, </span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">y =</span> Max<span class="sc">+</span><span class="fu">max</span>(Max)<span class="sc">*</span><span class="fl">0.1</span>, </span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">label =</span> groups))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="BigRedStatisticsInR_files/figure-html/unnamed-chunk-74-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="appendix-3.-create-dataframes-in-r" class="level2" data-link="How to graph in R">
<h2 data-link="How to graph in R" class="anchored" data-anchor-id="appendix-3.-create-dataframes-in-r">Appendix 3. Create dataframes in R</h2>
<p>Sometime you may need to create a dataframe in R for testing purpose.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a dataframe called fertilizer2</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>fertilizer2 <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="st">"Treatment"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"Soil"</span>, <span class="dv">5</span>), </span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>                                        <span class="fu">rep</span>(<span class="st">"Soil+fertilizer"</span>, <span class="dv">5</span>)),</span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Plant_height_cm"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="fl">10.5</span>, <span class="dv">13</span>, <span class="dv">9</span>, <span class="dv">11</span>, <span class="dv">14</span>, </span>
<span id="cb92-5"><a href="#cb92-5" aria-hidden="true" tabindex="-1"></a>                                              <span class="dv">17</span>, <span class="fl">18.5</span>, <span class="dv">20</span>, <span class="dv">16</span>, <span class="fl">17.5</span>))</span>
<span id="cb92-6"><a href="#cb92-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-7"><a href="#cb92-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Note that the new lines are only used for a better display. They are not interpreted by R.</span></span>
<span id="cb92-8"><a href="#cb92-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-9"><a href="#cb92-9" aria-hidden="true" tabindex="-1"></a><span class="co"># rep("Soil", 5) repeats "Soil" five times. It's the same as c("Soil", "Soil", "Soil", "Soil", "Soil")</span></span>
<span id="cb92-10"><a href="#cb92-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-11"><a href="#cb92-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-12"><a href="#cb92-12" aria-hidden="true" tabindex="-1"></a><span class="co"># You can save the dataframe as a csv file:</span></span>
<span id="cb92-13"><a href="#cb92-13" aria-hidden="true" tabindex="-1"></a><span class="co"># write.csv(fertilizer2, "fertilizer2.csv", row.names = F)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>